<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.475">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Case study on Kang18 scRNA-seq dataset</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="01-analysis-Kang18_files/libs/clipboard/clipboard.min.js"></script>
<script src="01-analysis-Kang18_files/libs/quarto-html/quarto.js"></script>
<script src="01-analysis-Kang18_files/libs/quarto-html/popper.min.js"></script>
<script src="01-analysis-Kang18_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="01-analysis-Kang18_files/libs/quarto-html/anchor.min.js"></script>
<link href="01-analysis-Kang18_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01-analysis-Kang18_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="01-analysis-Kang18_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="01-analysis-Kang18_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="01-analysis-Kang18_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#dependencies" id="toc-dependencies" class="nav-link active" data-scroll-target="#dependencies">Dependencies</a></li>
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a></li>
  <li><a href="#preprocessing-with-hvg" id="toc-preprocessing-with-hvg" class="nav-link" data-scroll-target="#preprocessing-with-hvg">Preprocessing with HVG</a></li>
  <li><a href="#initial-clustering-high-and-low-resolution" id="toc-initial-clustering-high-and-low-resolution" class="nav-link" data-scroll-target="#initial-clustering-high-and-low-resolution">Initial clustering: high and low resolution</a></li>
  <li><a href="#feature-scoring" id="toc-feature-scoring" class="nav-link" data-scroll-target="#feature-scoring">Feature scoring</a>
  <ul class="collapse">
  <li><a href="#further-exploration-of-different-scores" id="toc-further-exploration-of-different-scores" class="nav-link" data-scroll-target="#further-exploration-of-different-scores">Further exploration of different scores</a></li>
  <li><a href="#determine-the-optimal-size-of-features" id="toc-determine-the-optimal-size-of-features" class="nav-link" data-scroll-target="#determine-the-optimal-size-of-features">Determine the optimal size of features</a></li>
  </ul></li>
  <li><a href="#feature-selection-and-reprosessing" id="toc-feature-selection-and-reprosessing" class="nav-link" data-scroll-target="#feature-selection-and-reprosessing">Feature selection and reprosessing</a>
  <ul class="collapse">
  <li><a href="#feature-selection-step" id="toc-feature-selection-step" class="nav-link" data-scroll-target="#feature-selection-step">Feature selection step</a></li>
  <li><a href="#pca-and-umap-on-reprocessed-data" id="toc-pca-and-umap-on-reprocessed-data" class="nav-link" data-scroll-target="#pca-and-umap-on-reprocessed-data">PCA and UMAP on reprocessed data</a></li>
  <li><a href="#abundance-plot" id="toc-abundance-plot" class="nav-link" data-scroll-target="#abundance-plot">Abundance plot</a></li>
  </ul></li>
  <li><a href="#differential-analysis" id="toc-differential-analysis" class="nav-link" data-scroll-target="#differential-analysis">Differential analysis</a>
  <ul class="collapse">
  <li><a href="#da" id="toc-da" class="nav-link" data-scroll-target="#da">DA</a></li>
  <li><a href="#ds" id="toc-ds" class="nav-link" data-scroll-target="#ds">DS</a></li>
  <li><a href="#comparison-between-da-and-ds" id="toc-comparison-between-da-and-ds" class="nav-link" data-scroll-target="#comparison-between-da-and-ds">Comparison between DA and DS</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Case study on Kang18 scRNA-seq dataset</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="dependencies" class="level2">
<h2 class="anchored" data-anchor-id="dependencies">Dependencies</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scater)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ExperimentHub)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(limma)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(leiden)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(igraph)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(harmony)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">source</span>(<span class="st">"utils.R"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(edgeR)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scuttle)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(MLVSBM)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(GGally)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(miloR)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(miloDE)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>    <span class="co">#library(lemur)</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(stringr)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(UpSetR)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pathviewr)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(cluster)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(bluster)</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(viridis)</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(leiden)</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggridges)</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(FEAST)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(DUBStepR)</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load data</h2>
<p>Load filtered data (after QC and normalization) and subsample 3000 genes and 5000 cells for analysis.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load preprocessed data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/case/Kang18.rds"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>gid <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">nrow</span>(sce)), <span class="dv">3000</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>cid <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">ncol</span>(sce)), <span class="dv">5000</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[gid, cid]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="preprocessing-with-hvg" class="level2">
<h2 class="anchored" data-anchor-id="preprocessing-with-hvg">Preprocessing with HVG</h2>
<p>We selected highly variable genes (HVG) and use them for dimension reduction (DR). The DR result is shown in PCA colored by the annotated cell types, sample ids and conditions (control or stimulated).</p>
<p>First, we need to identify significant PCs in order to capture the majority of the variation between clusters or cell types. An elbow plot is drawn to show the percentage of explained variance by each PC using the top 50 PCs.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># widely used feature selection: HVG</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>tbl <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce, <span class="at">block =</span> sce<span class="sc">$</span>sample_id)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg <span class="ot">&lt;-</span> hvg <span class="ot">&lt;-</span> tbl<span class="sc">$</span>bio <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>bio <span class="ot">&lt;-</span> tbl<span class="sc">$</span>bio </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> hvg, <span class="at">ncomponent =</span> <span class="dv">50</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>pcs <span class="ot">&lt;-</span> <span class="fu">attr</span>(<span class="fu">reducedDim</span>(sce, <span class="st">"PCA"</span>), <span class="st">"percentVar"</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>pv <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">row.names =</span> <span class="cn">NULL</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC =</span> <span class="fu">seq_len</span>(<span class="fu">length</span>(pcs)),</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">percentVar =</span> pcs)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pv, <span class="fu">aes</span>(<span class="at">x =</span> PC, <span class="at">y =</span> percentVar, <span class="at">label =</span> PC)) <span class="sc">+</span> </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">size =</span> <span class="dv">3</span>, <span class="at">color =</span> <span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/Preprocessing-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
<p>With the elbow plot, we can roughly see the explained variance . To extract the information in a more quantitative manner, function is used to determine the exact elbow.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> hvg, <span class="at">ncomponents =</span> <span class="dv">10</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co"># integration</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">HarmonyMatrix</span>(<span class="at">data_mat =</span> <span class="fu">assay</span>(sce, <span class="st">"counts"</span>),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">meta_data =</span> <span class="fu">colData</span>(sce),</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">vars_use =</span> <span class="st">"sample_id"</span>,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">npcs =</span> <span class="dv">10</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(sce, <span class="st">"PCA"</span>) <span class="ot">&lt;-</span> pca</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true" href="">PCA - cell type</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false" href="">PCA - sample_id</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false" href="">PCA - group_id</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"cluster_id"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/PCA%20-%20cell%20type-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"sample_id"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/PCA%20-%20sample_id-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"group_id"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/PCA%20-%20group_id-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="initial-clustering-high-and-low-resolution" class="level2">
<h2 class="anchored" data-anchor-id="initial-clustering-high-and-low-resolution">Initial clustering: high and low resolution</h2>
<p>We first perform initial clustering in both high and low resolution using HVG. High resolution is for calculating type scores; while low resolution is for calculating state scores.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rsl &lt;- seq(0.1, 5, by = 0.2)</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># g &lt;- buildSNNGraph(sce, use.dimred = "PCA")</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># reso &lt;- lapply(rsl, \(i) {</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#     sce$cluster_hi &lt;- cluster_louvain(g, resolution = i)$membership</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#     sce$cluster_hi  &lt;- factor(sce$cluster_hi,sort(unique(sce$cluster_hi)))</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="co">#     sce$cluster_lo &lt;- sce$cluster_hi</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co">#     Fstat &lt;- .Fstat(sce)</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co">#     edgeR &lt;- .edgeR(sce)</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co">#     Fstat_edgeR &lt;- rank(Fstat) - rank(edgeR)</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co">#     data.frame(resolution = i, Fstat = Fstat, edgeR = edgeR, Fstat_edgeR)</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co"># })</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="co"># res_df &lt;- data.frame(do.call(rbind, reso))</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="co"># saveRDS(reso, "../data/case/reso.rds")</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>reso <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/case/reso.rds"</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, reso)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true" href="">Fstat</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false" href="">edgeR</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-3" role="tab" aria-controls="tabset-2-3" aria-selected="false" href="">Fstat_edgeR</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> Fstat, <span class="at">y =</span> <span class="fu">factor</span>(resolution), </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>() <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"clustering resolution"</span>) <span class="sc">+</span> </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Fstat vs. clustering resolution"</span>) <span class="sc">+</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(resolution), <span class="at">y =</span> Fstat, </span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span> </span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.07</span>) <span class="sc">+</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"resolution"</span>)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0793</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/Fstat-1.svg" class="img-fluid" width="1440"></p>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> res_df[res_df<span class="sc">$</span>edgeR <span class="sc">!=</span> <span class="dv">0</span>, ]</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">na.omit</span>(res_df), <span class="fu">aes</span>(<span class="at">x =</span> edgeR, <span class="at">y =</span> <span class="fu">factor</span>(resolution), </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>() <span class="sc">+</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"clustering resolution"</span>) <span class="sc">+</span> </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"edgeR -log(p_val) vs. clustering resolution"</span>) <span class="sc">+</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(resolution), <span class="at">y =</span> edgeR, </span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span> </span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.07</span>) <span class="sc">+</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"resolution"</span>) <span class="sc">+</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"edgeR -log(p_val)"</span>)</span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>p3 <span class="sc">+</span> p4 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.0393</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/edgeR-1.svg" class="img-fluid" width="1440"></p>
</div>
</div>
</div>
<div id="tabset-2-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-3-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> res_df[res_df<span class="sc">$</span>Fstat_edgeR <span class="sc">!=</span> <span class="dv">0</span>, ]</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">na.omit</span>(res_df), <span class="fu">aes</span>(<span class="at">x =</span> Fstat_edgeR, <span class="at">y =</span> <span class="fu">factor</span>(resolution), </span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>() <span class="sc">+</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"clustering resolution"</span>) <span class="sc">+</span> </span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Fstat_edgeR vs. clustering resolution"</span>) <span class="sc">+</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(resolution), <span class="at">y =</span> Fstat_edgeR, </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span> </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.07</span>) <span class="sc">+</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"resolution"</span>) <span class="sc">+</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Fstat_edgeR"</span>)</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>p5 <span class="sc">+</span> p6 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 0.1</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/Fstat_edgeR-1.svg" class="img-fluid" width="1440"></p>
</div>
</div>
</div>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(sce, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_hi <span class="ot">&lt;-</span> <span class="fu">cluster_louvain</span>(g, <span class="at">resolution =</span> <span class="dv">5</span>)<span class="sc">$</span>membership</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_lo <span class="ot">&lt;-</span> <span class="fu">cluster_louvain</span>(g, <span class="at">resolution =</span> <span class="fl">0.6</span>)<span class="sc">$</span>membership</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_hi <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce<span class="sc">$</span>cluster_hi, <span class="fu">sort</span>(<span class="fu">unique</span>(sce<span class="sc">$</span>cluster_hi)))</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_lo <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce<span class="sc">$</span>cluster_lo, <span class="fu">sort</span>(<span class="fu">unique</span>(sce<span class="sc">$</span>cluster_lo)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true" href="">PCA - hvg, high resolution</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false" href="">PCA - hvg, low resolution</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"cluster_hi"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/PCA%20high-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"cluster_lo"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/PCA%20low-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="feature-scoring" class="level2">
<h2 class="anchored" data-anchor-id="feature-scoring">Feature scoring</h2>
<p>Here, we calculated two type scores (F-statistics and entropy), one state scores (p values of DS analysis by edgeR) and random score drawed from uniform distribution ranging from 0 to 1. The following figure illustrates the relationship between different scores colored by HVG. We also added two published feature selection tools into comparison: <em>FEAST</em> which is also a clustering-based filter method scoring the genes by F-statistics, and <em>DUBStepR</em> that is based on the idea of cell-type-specific markers should be well correlated with each other.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat <span class="ot">&lt;-</span> <span class="fu">.Fstat</span>(sce)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR <span class="ot">&lt;-</span> <span class="fu">.edgeR</span>(sce)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>random <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="fu">nrow</span>(sce))</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR <span class="ot">&lt;-</span> <span class="fu">rank</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat) <span class="sc">-</span> <span class="fu">rank</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST <span class="ot">&lt;-</span> <span class="fu">.FEAST</span>(sce)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR <span class="ot">&lt;-</span> <span class="fu">.DUBStepR</span>(sce)</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">rowData</span>(sce)) <span class="sc">%&gt;%</span> </span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">Fstat =</span> <span class="fu">log10</span>(Fstat),</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">edgeR =</span> <span class="fu">log10</span>(edgeR))</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(df, <span class="at">columns =</span> <span class="dv">6</span><span class="sc">:</span><span class="dv">11</span>, </span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>    ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">colour =</span> hvg, <span class="at">alpha =</span> <span class="fl">0.4</span>),</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">upper =</span> <span class="st">"blank"</span>,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/Score%20features-1.svg" class="img-fluid" width="864"></p>
</div>
</div>
<section id="further-exploration-of-different-scores" class="level3">
<h3 class="anchored" data-anchor-id="further-exploration-of-different-scores">Further exploration of different scores</h3>
<p>We utilized the <code>isOutlier</code> function to identify genes whose type/state scores were outliers, using the median absolute deviation (MAD) as the criterion. For highly variable genes (HVG), we selected those with a biological variation greater than 0. Furthermore, <strong>DUBStepR</strong> has already furnished an optimal set of features. An ideal marker is expected to exhibit a high Fstat score but a low edgeR score, signifying a high type score yet a low state score. To ensure each method selected a sufficient number of genes, we set the <code>nmads</code> parameter to 1.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="dv">1</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="dv">1</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="fl">0.1</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg_sel <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>bio <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="dv">1</span>)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Fstat =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_sel,<span class="st">"gene_id"</span>], </span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">Fstat_edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">hvg =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">FEAST =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">DUBStepR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel,<span class="st">"gene_id"</span>])</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(lst),</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>        <span class="at">order.by =</span> <span class="st">"freq"</span>,</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>        <span class="at">nsets =</span> <span class="fu">length</span>(lst),</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>        <span class="at">set_size.show =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">scale.sets =</span> <span class="st">"identity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/unnamed-chunk-7-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
<p>Here, we extract genes with rank difference between Fstat type score and edgeR state score larger than 0, indicating a gene has relatively larger type score than state score. It is observed that even genes with every high F-statistics values but high edgeR value might still be excluded by <em>Fstat_edgeR</em> selection.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(sel), <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(Fstat), <span class="at">y =</span> edgeR, <span class="at">col =</span> Fstat_edgeR)) <span class="sc">+</span> </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradient2</span>() <span class="sc">+</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"-log(p_val) in edgeR DS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/unnamed-chunk-8-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="determine-the-optimal-size-of-features" class="level3">
<h3 class="anchored" data-anchor-id="determine-the-optimal-size-of-features">Determine the optimal size of features</h3>
<p>To determine the optimal size of selected features, we tried different number of genes and evaluate its performance using ARI, mean silhouette score and neighbour purity between clusters. There are two dashed lines: the black dashed line indicates the evaluation value when no feature selection, i.e.&nbsp;using all features; the red dashed line shows the results by <strong>DUBStepR</strong> because it provides an optimal set of features and thus no need to vary the feature size.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR"</span>, <span class="st">"random"</span>, <span class="st">"bio"</span>, <span class="st">"Fstat"</span>, <span class="st">"FEAST"</span>)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fl">0.1</span>,<span class="dv">1</span>,<span class="fl">0.05</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>params <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">n =</span> n, <span class="at">sel =</span> sel)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over each row of params and compute performance</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(params), \(i) {</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> params<span class="sc">$</span>n[i]</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  s <span class="ot">&lt;-</span> params<span class="sc">$</span>sel[i]</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowData</span>(sce)[, s], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="fu">seq_len</span>(<span class="fu">round</span>(<span class="fu">nrow</span>(sce)<span class="sc">*</span>x))]</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>  sel_val <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">nrow</span>(sce))</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>  sel_val[idx] <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> sel_val)</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>  sce<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>  ari <span class="ot">&lt;-</span> <span class="fu">ARI</span>(sce<span class="sc">$</span>cluster_re, sce<span class="sc">$</span>cluster_id)</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>  sil <span class="ot">&lt;-</span> <span class="fu">.sil_k</span>(sce)</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>  pur <span class="ot">&lt;-</span> <span class="fu">.pur_k</span>(sce)</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">ARI =</span> ari, <span class="at">sil =</span> sil, <span class="at">pur =</span> pur, <span class="at">sel =</span> s, <span class="at">n =</span> x)</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>sta <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">do.call</span>(rbind, res))</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="co">#saveRDS(sta, "../data/case/sta.rds")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co">#sta &lt;- readRDS("/Volumes/jiayiwang/type-state/data/case/sta.rds")</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ARI =</span> sta<span class="sc">$</span>ARI,</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">precentage =</span> n,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">silhouette =</span> sta<span class="sc">$</span>sil,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">purity =</span> sta<span class="sc">$</span>pur,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">number =</span> <span class="fu">nrow</span>(sce)<span class="sc">*</span>n,</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> sta<span class="sc">$</span>sel) <span class="sc">%&gt;%</span> </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">paste0</span>(number, <span class="st">", "</span>, <span class="fu">round</span>(ARI, <span class="at">digits =</span> <span class="dv">2</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">sel =</span> <span class="fu">case_when</span>(sel <span class="sc">==</span> <span class="st">"bio"</span> <span class="sc">~</span> <span class="st">"hvg"</span>,</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> sel))</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="co"># baseline</span></span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>all <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">TRUE</span>, <span class="fu">nrow</span>(sce))</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>sce1 <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> all)</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>sce1<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce1, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>base_ari <span class="ot">&lt;-</span> <span class="fu">ARI</span>(sce1<span class="sc">$</span>cluster_re, sce1<span class="sc">$</span>cluster_id)</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>base_sil <span class="ot">&lt;-</span> <span class="fu">.sil_k</span>(sce1)</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>base_pur <span class="ot">&lt;-</span> <span class="fu">.pur_k</span>(sce1)</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="co"># DUBStepR</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel)</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>sce2<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce2, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>dub_ari <span class="ot">&lt;-</span> <span class="fu">ARI</span>(sce2<span class="sc">$</span>cluster_re, sce2<span class="sc">$</span>cluster_id)</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>dub_sil <span class="ot">&lt;-</span> <span class="fu">.sil_k</span>(sce2)</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a>dub_pur <span class="ot">&lt;-</span> <span class="fu">.pur_k</span>(sce2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true" href="">ARI</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false" href="">Silhouette Score</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-3" role="tab" aria-controls="tabset-4-3" aria-selected="false" href="">Neighbourhood purity</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> precentage, <span class="at">y =</span> ARI, <span class="at">col =</span> sel)) <span class="sc">+</span> </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> base_ari, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> dub_ari, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">0.1</span>, <span class="at">y =</span> base_ari <span class="sc">+</span> <span class="fl">0.02</span>, </span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">label =</span> <span class="st">"no feature selection"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"ARI vs. precentage of the selected features"</span>) <span class="sc">+</span></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/unnamed-chunk-11-1.svg" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> precentage, <span class="at">y =</span> silhouette, <span class="at">col =</span> sel)) <span class="sc">+</span> </span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> base_sil, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> dub_sil, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>     <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">0.1</span>, <span class="at">y =</span> base_sil <span class="sc">+</span> <span class="fl">0.01</span>, </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">label =</span> <span class="st">"no feature selection"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"silhouette vs. precentage of the selected features"</span>) <span class="sc">+</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/unnamed-chunk-12-1.svg" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-4-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-3-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> precentage, <span class="at">y =</span> purity, <span class="at">col =</span> sel)) <span class="sc">+</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> base_pur, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> dub_pur, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>     <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">0.1</span>, <span class="at">y =</span> base_pur <span class="sc">-</span> <span class="fl">0.03</span>, </span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">label =</span> <span class="st">"no feature selection"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"NeighbourPurity vs. precentage of the selected features"</span>) <span class="sc">+</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/unnamed-chunk-13-1.svg" class="img-fluid" width="768"></p>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="feature-selection-and-reprosessing" class="level2">
<h2 class="anchored" data-anchor-id="feature-selection-and-reprosessing">Feature selection and reprosessing</h2>
<section id="feature-selection-step" class="level3">
<h3 class="anchored" data-anchor-id="feature-selection-step">Feature selection step</h3>
<p>The above evaluation statistics results suggested that different selection methods have their own specific optimal size. Now we simply selected top 40% genes based on their scores (except for hvg and <strong>DUBStepR</strong>), which requires further adaption of the feature sizes.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>fs <span class="ot">&lt;-</span> \(sce, n, sel, res) {</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    idx <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowData</span>(sce)[,sel], </span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="fu">seq_len</span>(<span class="fu">round</span>(<span class="fu">nrow</span>(sce)<span class="sc">*</span>n))]</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>    sel_val <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">nrow</span>(sce))</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>    sel_val[idx] <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rowData</span>(sce)[, res] <span class="ot">&lt;-</span> sel_val</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(sce)</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>sco <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR"</span>, <span class="st">"Fstat"</span>, <span class="st">"random"</span>, <span class="st">"FEAST"</span>)</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (s <span class="cf">in</span> sel) {</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>  r <span class="ot">&lt;-</span> <span class="fu">paste0</span>(s,<span class="st">"_sel"</span>)</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">fs</span>(sce, <span class="at">n =</span> <span class="fl">0.4</span>, <span class="at">sel =</span> s, <span class="at">res =</span> r)</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR_sel"</span>, <span class="st">"Fstat_sel"</span>, <span class="st">"random_sel"</span>, </span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"hvg"</span>, <span class="st">"DUBStepR_sel"</span>, <span class="st">"FEAST_sel"</span>)</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>sces <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sel, \(s) {</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> <span class="fu">rowData</span>(sce)[,s])</span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>  sce<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce)</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>}) </span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sces) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR"</span>, <span class="st">"Fstat"</span>, <span class="st">"random"</span>, <span class="st">"hvg"</span>, <span class="st">"DUBStepR"</span>, <span class="st">"FEAST"</span>)</span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Fstat =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_sel,<span class="st">"gene_id"</span>], </span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a>  <span class="at">Fstat_edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">hvg =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">FEAST =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb27-30"><a href="#cb27-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">DUBStepR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel,<span class="st">"gene_id"</span>])</span>
<span id="cb27-31"><a href="#cb27-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-32"><a href="#cb27-32" aria-hidden="true" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(lst),</span>
<span id="cb27-33"><a href="#cb27-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">order.by =</span> <span class="st">"freq"</span>,</span>
<span id="cb27-34"><a href="#cb27-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">nsets =</span> <span class="fu">length</span>(lst),</span>
<span id="cb27-35"><a href="#cb27-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">set_size.show =</span> <span class="cn">TRUE</span>,</span>
<span id="cb27-36"><a href="#cb27-36" aria-hidden="true" tabindex="-1"></a>        <span class="at">scale.sets =</span> <span class="st">"identity"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/Feature%20selection%20and%20reprocess-1.svg" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="pca-and-umap-on-reprocessed-data" class="level3">
<h3 class="anchored" data-anchor-id="pca-and-umap-on-reprocessed-data">PCA and UMAP on reprocessed data</h3>
<p>PCA plots on selected features indicated that all other methods are affected by the strong state effect in the data, with two conditions clearly separated. However, our Fstat_edgeR gives two blobs with two conditions are well mixed.</p>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true" href="">new PCA, cluster_re</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-2" role="tab" aria-controls="tabset-5-2" aria-selected="false" href="">new PCA, condition</a></li></ul>
<div class="tab-content">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>pk <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) <span class="fu">plotPCA</span>(sces[[x]], <span class="at">colour_by =</span> <span class="st">"cluster_re"</span>) <span class="sc">+</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(x)) <span class="sc">|&gt;</span> </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">wrap_plots</span>(<span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>pk</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/pk-1.svg" class="img-fluid" width="768"></p>
</div>
</div>
</div>
<div id="tabset-5-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>pg <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) <span class="fu">plotPCA</span>(sces[[x]], <span class="at">colour_by =</span> <span class="st">"group_id"</span>) <span class="sc">+</span> </span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(x)) <span class="sc">|&gt;</span> </span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">wrap_plots</span>(<span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>pg</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/pg-1.svg" class="img-fluid" width="768"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="abundance-plot" class="level3">
<h3 class="anchored" data-anchor-id="abundance-plot">Abundance plot</h3>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true" href="">Proportion of cell types in each cluster_re</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-2" role="tab" aria-controls="tabset-6-2" aria-selected="false" href="">Proportion of cluster_re in each cell type</a></li></ul>
<div class="tab-content">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x){</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    cd <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(sces[[x]]))</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(cd) <span class="sc">+</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">fill =</span> cluster_id) <span class="sc">+</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">xlab</span>(<span class="st">"Cluster after reprocessed"</span>) <span class="sc">+</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>) <span class="sc">+</span> </span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggtitle</span>(x)</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">wrap_plots</span>(<span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/Abundance-1.svg" class="img-fluid" width="1152"></p>
</div>
</div>
</div>
<div id="tabset-6-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x){</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    cd <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(sces[[x]]))</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(cd) <span class="sc">+</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> cluster_id, <span class="at">fill =</span> cluster_re) <span class="sc">+</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>        <span class="fu">xlab</span>(<span class="st">"Cell Type"</span>) <span class="sc">+</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>) <span class="sc">+</span> </span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggtitle</span>(x) <span class="sc">+</span> </span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">wrap_plots</span>(<span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/abundance%20plot2-1.svg" class="img-fluid" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="differential-analysis" class="level2">
<h2 class="anchored" data-anchor-id="differential-analysis">Differential analysis</h2>
<section id="da" class="level3">
<h3 class="anchored" data-anchor-id="da">DA</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># edgeR</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>dal <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) {</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  e <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DA_edgeR</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, <span class="st">"cluster_re"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DA_edgeR"</span>, <span class="at">DADS =</span> <span class="st">"DA"</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>) </span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DA_limma</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, <span class="st">"cluster_re"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DA_limma"</span>, <span class="at">DADS =</span> <span class="st">"DA"</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DA_milo</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, <span class="st">"Nhood"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DA_milo"</span>, <span class="at">DADS =</span> <span class="st">"DA"</span>, <span class="at">clustering =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="st">"cluster_re"</span> <span class="ot">=</span> <span class="st">"Nhood"</span>)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(rbind, <span class="fu">list</span>(e,l,m))</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a>da <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, dal)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="ds" class="level3">
<h3 class="anchored" data-anchor-id="ds">DS</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>dsl <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) {</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  e <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DS_edgeR</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, </span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"cluster_re"</span>, <span class="st">"gene"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DS_edgeR"</span>, <span class="at">DADS =</span> <span class="st">"DS"</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>) </span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#l &lt;- data.frame(.DS_lemur(sces[[x]])[,c("p_val", "p_adj", "cluster_re")], </span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  sel = x, das = "DS_lemur", DADS = "DS", clustering = FALSE)</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DS_miloDE</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, </span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Nhood"</span>, <span class="st">"gene"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DS_miloDE"</span>, <span class="at">DADS =</span> <span class="st">"DS"</span>, <span class="at">clustering =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb33-10"><a href="#cb33-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="st">"cluster_re"</span> <span class="ot">=</span> <span class="st">"Nhood"</span>)</span>
<span id="cb33-11"><a href="#cb33-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(rbind, <span class="fu">list</span>(e,m))</span>
<span id="cb33-12"><a href="#cb33-12" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>100% covered by 13 sets.
100% covered by 23 sets.
100% covered by 17 sets.
100% covered by 21 sets.
100% covered by 23 sets.
100% covered by 16 sets.</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, dsl)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>das <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">list</span>(ds[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(ds) <span class="sc">==</span> <span class="st">"gene"</span>)], da))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="comparison-between-da-and-ds" class="level3">
<h3 class="anchored" data-anchor-id="comparison-between-da-and-ds">Comparison between DA and DS</h3>
<p>The DA and DS p-value distribution with different feature selection methods.</p>
<section id="distribution-of-p-values-and-logfc" class="level4">
<h4 class="anchored" data-anchor-id="distribution-of-p-values-and-logfc">Distribution of p values and logFC</h4>
<div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-7-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-1" role="tab" aria-controls="tabset-7-1" aria-selected="true" href="">By selection</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-2" role="tab" aria-controls="tabset-7-2" aria-selected="false" href="">By DAS</a></li></ul>
<div class="tab-content">
<div id="tabset-7-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-7-1-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(p_val, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> das, <span class="at">linetype =</span> DADS)) <span class="sc">+</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/sel-1.svg" class="img-fluid" width="672"></p>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(logFC, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> das, <span class="at">linetype =</span> DADS)) <span class="sc">+</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/sel-2.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
<div id="tabset-7-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-2-tab">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(p_val, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> sel)) <span class="sc">+</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> das, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/DAS-1.svg" class="img-fluid" width="672"></p>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(logFC, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> sel)) <span class="sc">+</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> das, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/DAS-2.svg" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="clustering-based-da-vs-clustering-based-ds" class="level4">
<h4 class="anchored" data-anchor-id="clustering-based-da-vs-clustering-based-ds">Clustering-based DA vs Clustering-based DS</h4>
<p>Since the number of clusters between clustering-based DA and DS are the same, we are able to compare the results of <em>DA_edgeR</em>, <em>DA_limma</em> with the aggregated results of <em>DS_edgeR</em> in each cluster.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> das[das<span class="sc">$</span>clustering <span class="sc">==</span> <span class="cn">TRUE</span>,]</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">color =</span> <span class="fu">case_when</span>(</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"pink"</span>,</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">~</span> <span class="st">"lightblue"</span>,</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>    DADS <span class="sc">==</span> <span class="st">"DA"</span> <span class="sc">~</span> <span class="st">"black"</span>,</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"gray"</span></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> p_val, <span class="at">col =</span> DADS, <span class="at">shape =</span> das)) <span class="sc">+</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Pastel1"</span>) <span class="sc">+</span></span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> <span class="st">"mean"</span>, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.05</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/DADS-1.svg" class="img-fluid" width="1152"></p>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> logFC, <span class="at">shape =</span> das, <span class="at">col =</span> color)) <span class="sc">+</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_identity</span>() <span class="sc">+</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> <span class="st">"mean"</span>, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">8</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/DADS-2.svg" class="img-fluid" width="1152"></p>
</div>
</div>
</section>
<section id="clustering-based-da-vs-clustering-free-da" class="level4">
<h4 class="anchored" data-anchor-id="clustering-based-da-vs-clustering-free-da">Clustering-based DA vs Clustering-free DA</h4>
<p>In the <code>miloR</code> package, the <code>annotateNhood</code> function can be used to annotate each neighborhood with a cluster ID. This is achieved by calculating the maximum proportion of clusters present in that neighborhood.</p>
<p>The beeswarm plots are then generated to compare the mean logFC or p-val values over neighborhoods in each cluster with the clustering-based results. These plots provide a visual comparison between the clustering-based results and the mean logFC or p-val values for each neighborhood within the respective clusters.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>milos <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(i){</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> sces[[i]]</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>    cd <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(x))</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(cd[<span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"group_id"</span>)])</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(df) <span class="ot">&lt;-</span> df<span class="sc">$</span>sample_id</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">#x &lt;- Milo(x)</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">makeNhoods</span>(<span class="fu">buildGraph</span>(<span class="fu">Milo</span>(x), <span class="at">d =</span> <span class="fu">ncol</span>(<span class="fu">reducedDim</span>(x, <span class="st">"PCA"</span>))))</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">countCells</span>(y, <span class="st">"sample_id"</span>, cd)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">calcNhoodDistance</span>(y, <span class="at">d =</span> <span class="fu">ncol</span>(<span class="fu">reducedDim</span>(x, <span class="st">"PCA"</span>)))</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>    da <span class="ot">&lt;-</span> <span class="fu">testNhoods</span>(y, <span class="sc">~</span> group_id, df)</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">annotateNhoods</span>(y, <span class="at">da.res =</span> da, <span class="at">coldata_col =</span> <span class="st">"cluster_re"</span>)</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(res, <span class="at">sel =</span> i)</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, milos)</span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> df1 <span class="sc">%&gt;%</span></span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">color =</span> <span class="fu">case_when</span>(</span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"pink"</span>,</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">~</span> <span class="st">"lightblue"</span>,</span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"gray"</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> das[das<span class="sc">$</span>clustering <span class="sc">==</span> <span class="cn">TRUE</span> <span class="sc">&amp;</span> das<span class="sc">$</span>DADS <span class="sc">==</span> <span class="st">"DA"</span>,]</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df1, <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> logFC, <span class="at">col =</span> color)) <span class="sc">+</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_identity</span>() <span class="sc">+</span></span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> <span class="st">"mean"</span>, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span> </span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df2, </span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a>             <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> logFC, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">shape =</span> das))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="01-analysis-Kang18_files/figure-html/milo%20comparsion-1.svg" class="img-fluid" width="1152"></p>
</div>
</div>
<!-- -->

</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb43" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Case study on Kang18 scRNA-seq dataset"</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: true</span></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a><span class="co">    toc_float: true</span></span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a><span class="fu">## Dependencies </span></span>
<span id="cb43-13"><a href="#cb43-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-14"><a href="#cb43-14" aria-hidden="true" tabindex="-1"></a><span class="in">```{r warning=FALSE}</span></span>
<span id="cb43-15"><a href="#cb43-15" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-libs</span></span>
<span id="cb43-16"><a href="#cb43-16" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb43-17"><a href="#cb43-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scran)</span>
<span id="cb43-18"><a href="#cb43-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scater)</span>
<span id="cb43-19"><a href="#cb43-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(patchwork)</span>
<span id="cb43-20"><a href="#cb43-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ExperimentHub)</span>
<span id="cb43-21"><a href="#cb43-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(limma)</span>
<span id="cb43-22"><a href="#cb43-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(leiden)</span>
<span id="cb43-23"><a href="#cb43-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(igraph)</span>
<span id="cb43-24"><a href="#cb43-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(harmony)</span>
<span id="cb43-25"><a href="#cb43-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">source</span>(<span class="st">"utils.R"</span>)</span>
<span id="cb43-26"><a href="#cb43-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(dplyr)</span>
<span id="cb43-27"><a href="#cb43-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(edgeR)</span>
<span id="cb43-28"><a href="#cb43-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(scuttle)</span>
<span id="cb43-29"><a href="#cb43-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(MLVSBM)</span>
<span id="cb43-30"><a href="#cb43-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(GGally)</span>
<span id="cb43-31"><a href="#cb43-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(miloR)</span>
<span id="cb43-32"><a href="#cb43-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(miloDE)</span>
<span id="cb43-33"><a href="#cb43-33" aria-hidden="true" tabindex="-1"></a>    <span class="co">#library(lemur)</span></span>
<span id="cb43-34"><a href="#cb43-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(stringr)</span>
<span id="cb43-35"><a href="#cb43-35" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(UpSetR)</span>
<span id="cb43-36"><a href="#cb43-36" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(pathviewr)</span>
<span id="cb43-37"><a href="#cb43-37" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(cluster)</span>
<span id="cb43-38"><a href="#cb43-38" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(bluster)</span>
<span id="cb43-39"><a href="#cb43-39" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(viridis)</span>
<span id="cb43-40"><a href="#cb43-40" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(leiden)</span>
<span id="cb43-41"><a href="#cb43-41" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggbeeswarm)</span>
<span id="cb43-42"><a href="#cb43-42" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(ggridges)</span>
<span id="cb43-43"><a href="#cb43-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(FEAST)</span>
<span id="cb43-44"><a href="#cb43-44" aria-hidden="true" tabindex="-1"></a>    <span class="fu">library</span>(DUBStepR)</span>
<span id="cb43-45"><a href="#cb43-45" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-46"><a href="#cb43-46" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb43-47"><a href="#cb43-47" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-48"><a href="#cb43-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-49"><a href="#cb43-49" aria-hidden="true" tabindex="-1"></a><span class="fu">## Load data</span></span>
<span id="cb43-50"><a href="#cb43-50" aria-hidden="true" tabindex="-1"></a>Load filtered data (after QC and normalization) and subsample 3000 genes and 5000 cells for analysis. </span>
<span id="cb43-51"><a href="#cb43-51" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fold}</span></span>
<span id="cb43-52"><a href="#cb43-52" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: load-data</span></span>
<span id="cb43-53"><a href="#cb43-53" aria-hidden="true" tabindex="-1"></a><span class="co"># load preprocessed data</span></span>
<span id="cb43-54"><a href="#cb43-54" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/case/Kang18.rds"</span>)</span>
<span id="cb43-55"><a href="#cb43-55" aria-hidden="true" tabindex="-1"></a>gid <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">nrow</span>(sce)), <span class="dv">3000</span>)</span>
<span id="cb43-56"><a href="#cb43-56" aria-hidden="true" tabindex="-1"></a>cid <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="fu">seq_len</span>(<span class="fu">ncol</span>(sce)), <span class="dv">5000</span>)</span>
<span id="cb43-57"><a href="#cb43-57" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> sce[gid, cid]</span>
<span id="cb43-58"><a href="#cb43-58" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-59"><a href="#cb43-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-60"><a href="#cb43-60" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preprocessing with HVG</span></span>
<span id="cb43-61"><a href="#cb43-61" aria-hidden="true" tabindex="-1"></a>We selected highly variable genes (HVG) and use them for dimension reduction (DR). The DR result is shown in PCA colored by the annotated cell types, sample ids and conditions (control or stimulated). </span>
<span id="cb43-62"><a href="#cb43-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-63"><a href="#cb43-63" aria-hidden="true" tabindex="-1"></a>First, we need to identify significant PCs in order to capture the majority of the variation between clusters or cell types. An elbow plot is drawn to show the percentage of explained variance by each PC using the top 50 PCs. </span>
<span id="cb43-64"><a href="#cb43-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-65"><a href="#cb43-65" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fold}</span></span>
<span id="cb43-66"><a href="#cb43-66" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Preprocessing</span></span>
<span id="cb43-67"><a href="#cb43-67" aria-hidden="true" tabindex="-1"></a><span class="co"># widely used feature selection: HVG</span></span>
<span id="cb43-68"><a href="#cb43-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-69"><a href="#cb43-69" aria-hidden="true" tabindex="-1"></a>tbl <span class="ot">&lt;-</span> <span class="fu">modelGeneVar</span>(sce, <span class="at">block =</span> sce<span class="sc">$</span>sample_id)</span>
<span id="cb43-70"><a href="#cb43-70" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg <span class="ot">&lt;-</span> hvg <span class="ot">&lt;-</span> tbl<span class="sc">$</span>bio <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb43-71"><a href="#cb43-71" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>bio <span class="ot">&lt;-</span> tbl<span class="sc">$</span>bio </span>
<span id="cb43-72"><a href="#cb43-72" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> hvg, <span class="at">ncomponent =</span> <span class="dv">50</span>)</span>
<span id="cb43-73"><a href="#cb43-73" aria-hidden="true" tabindex="-1"></a>pcs <span class="ot">&lt;-</span> <span class="fu">attr</span>(<span class="fu">reducedDim</span>(sce, <span class="st">"PCA"</span>), <span class="st">"percentVar"</span>)</span>
<span id="cb43-74"><a href="#cb43-74" aria-hidden="true" tabindex="-1"></a>pv <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">row.names =</span> <span class="cn">NULL</span>,</span>
<span id="cb43-75"><a href="#cb43-75" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC =</span> <span class="fu">seq_len</span>(<span class="fu">length</span>(pcs)),</span>
<span id="cb43-76"><a href="#cb43-76" aria-hidden="true" tabindex="-1"></a>  <span class="at">percentVar =</span> pcs)</span>
<span id="cb43-77"><a href="#cb43-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-78"><a href="#cb43-78" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(pv, <span class="fu">aes</span>(<span class="at">x =</span> PC, <span class="at">y =</span> percentVar, <span class="at">label =</span> PC)) <span class="sc">+</span> </span>
<span id="cb43-79"><a href="#cb43-79" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">size =</span> <span class="dv">3</span>, <span class="at">color =</span> <span class="st">"blue"</span>)</span>
<span id="cb43-80"><a href="#cb43-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-81"><a href="#cb43-81" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-82"><a href="#cb43-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-83"><a href="#cb43-83" aria-hidden="true" tabindex="-1"></a>With the elbow plot, we can roughly see the explained variance . To extract the information in a more quantitative manner, \code{find_curve_elbow()} function is used to determine the exact elbow. </span>
<span id="cb43-84"><a href="#cb43-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-85"><a href="#cb43-85" aria-hidden="true" tabindex="-1"></a><span class="in">```{r harmony, message=FALSE, warning=FALSE}</span></span>
<span id="cb43-86"><a href="#cb43-86" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> hvg, <span class="at">ncomponents =</span> <span class="dv">10</span>)</span>
<span id="cb43-87"><a href="#cb43-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-88"><a href="#cb43-88" aria-hidden="true" tabindex="-1"></a><span class="co"># integration</span></span>
<span id="cb43-89"><a href="#cb43-89" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">HarmonyMatrix</span>(<span class="at">data_mat =</span> <span class="fu">assay</span>(sce, <span class="st">"counts"</span>),</span>
<span id="cb43-90"><a href="#cb43-90" aria-hidden="true" tabindex="-1"></a>    <span class="at">meta_data =</span> <span class="fu">colData</span>(sce),</span>
<span id="cb43-91"><a href="#cb43-91" aria-hidden="true" tabindex="-1"></a>    <span class="at">vars_use =</span> <span class="st">"sample_id"</span>,</span>
<span id="cb43-92"><a href="#cb43-92" aria-hidden="true" tabindex="-1"></a>    <span class="at">npcs =</span> <span class="dv">10</span>)</span>
<span id="cb43-93"><a href="#cb43-93" aria-hidden="true" tabindex="-1"></a><span class="fu">reducedDim</span>(sce, <span class="st">"PCA"</span>) <span class="ot">&lt;-</span> pca</span>
<span id="cb43-94"><a href="#cb43-94" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-95"><a href="#cb43-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-96"><a href="#cb43-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-97"><a href="#cb43-97" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb43-98"><a href="#cb43-98" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-99"><a href="#cb43-99" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA - cell type</span></span>
<span id="cb43-100"><a href="#cb43-100" aria-hidden="true" tabindex="-1"></a><span class="in">```{r PCA - cell type}</span></span>
<span id="cb43-101"><a href="#cb43-101" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"cluster_id"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span>
<span id="cb43-102"><a href="#cb43-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-103"><a href="#cb43-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-104"><a href="#cb43-104" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA - sample_id</span></span>
<span id="cb43-105"><a href="#cb43-105" aria-hidden="true" tabindex="-1"></a><span class="in">```{r PCA - sample_id}</span></span>
<span id="cb43-106"><a href="#cb43-106" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"sample_id"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span>
<span id="cb43-107"><a href="#cb43-107" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-108"><a href="#cb43-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-109"><a href="#cb43-109" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA - group_id</span></span>
<span id="cb43-110"><a href="#cb43-110" aria-hidden="true" tabindex="-1"></a><span class="in">```{r PCA - group_id}</span></span>
<span id="cb43-111"><a href="#cb43-111" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"group_id"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span>
<span id="cb43-112"><a href="#cb43-112" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-113"><a href="#cb43-113" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-114"><a href="#cb43-114" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-115"><a href="#cb43-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-116"><a href="#cb43-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-117"><a href="#cb43-117" aria-hidden="true" tabindex="-1"></a><span class="fu">## Initial clustering: high and low resolution</span></span>
<span id="cb43-118"><a href="#cb43-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-119"><a href="#cb43-119" aria-hidden="true" tabindex="-1"></a>We first perform initial clustering in both high and low resolution using HVG. High resolution is for calculating type scores; while low resolution is for calculating state scores. </span>
<span id="cb43-120"><a href="#cb43-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-121"><a href="#cb43-121" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.height=8, fig.width=15, warning=FALSE}</span></span>
<span id="cb43-122"><a href="#cb43-122" aria-hidden="true" tabindex="-1"></a><span class="co"># rsl &lt;- seq(0.1, 5, by = 0.2)</span></span>
<span id="cb43-123"><a href="#cb43-123" aria-hidden="true" tabindex="-1"></a><span class="co"># g &lt;- buildSNNGraph(sce, use.dimred = "PCA")</span></span>
<span id="cb43-124"><a href="#cb43-124" aria-hidden="true" tabindex="-1"></a><span class="co"># reso &lt;- lapply(rsl, \(i) {</span></span>
<span id="cb43-125"><a href="#cb43-125" aria-hidden="true" tabindex="-1"></a><span class="co">#     sce$cluster_hi &lt;- cluster_louvain(g, resolution = i)$membership</span></span>
<span id="cb43-126"><a href="#cb43-126" aria-hidden="true" tabindex="-1"></a><span class="co">#     sce$cluster_hi  &lt;- factor(sce$cluster_hi,sort(unique(sce$cluster_hi)))</span></span>
<span id="cb43-127"><a href="#cb43-127" aria-hidden="true" tabindex="-1"></a><span class="co">#     sce$cluster_lo &lt;- sce$cluster_hi</span></span>
<span id="cb43-128"><a href="#cb43-128" aria-hidden="true" tabindex="-1"></a><span class="co">#     Fstat &lt;- .Fstat(sce)</span></span>
<span id="cb43-129"><a href="#cb43-129" aria-hidden="true" tabindex="-1"></a><span class="co">#     edgeR &lt;- .edgeR(sce)</span></span>
<span id="cb43-130"><a href="#cb43-130" aria-hidden="true" tabindex="-1"></a><span class="co">#     Fstat_edgeR &lt;- rank(Fstat) - rank(edgeR)</span></span>
<span id="cb43-131"><a href="#cb43-131" aria-hidden="true" tabindex="-1"></a><span class="co">#     data.frame(resolution = i, Fstat = Fstat, edgeR = edgeR, Fstat_edgeR)</span></span>
<span id="cb43-132"><a href="#cb43-132" aria-hidden="true" tabindex="-1"></a><span class="co"># })</span></span>
<span id="cb43-133"><a href="#cb43-133" aria-hidden="true" tabindex="-1"></a><span class="co"># res_df &lt;- data.frame(do.call(rbind, reso))</span></span>
<span id="cb43-134"><a href="#cb43-134" aria-hidden="true" tabindex="-1"></a><span class="co"># saveRDS(reso, "../data/case/reso.rds")</span></span>
<span id="cb43-135"><a href="#cb43-135" aria-hidden="true" tabindex="-1"></a>reso <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/case/reso.rds"</span>)</span>
<span id="cb43-136"><a href="#cb43-136" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, reso)</span>
<span id="cb43-137"><a href="#cb43-137" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-138"><a href="#cb43-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-139"><a href="#cb43-139" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb43-140"><a href="#cb43-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-141"><a href="#cb43-141" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fstat</span></span>
<span id="cb43-142"><a href="#cb43-142" aria-hidden="true" tabindex="-1"></a><span class="in">```{r Fstat, fig.height=8, fig.width=15, warning=FALSE}</span></span>
<span id="cb43-143"><a href="#cb43-143" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> Fstat, <span class="at">y =</span> <span class="fu">factor</span>(resolution), </span>
<span id="cb43-144"><a href="#cb43-144" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span></span>
<span id="cb43-145"><a href="#cb43-145" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>() <span class="sc">+</span></span>
<span id="cb43-146"><a href="#cb43-146" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"clustering resolution"</span>) <span class="sc">+</span> </span>
<span id="cb43-147"><a href="#cb43-147" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb43-148"><a href="#cb43-148" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb43-149"><a href="#cb43-149" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Fstat vs. clustering resolution"</span>) <span class="sc">+</span></span>
<span id="cb43-150"><a href="#cb43-150" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>) </span>
<span id="cb43-151"><a href="#cb43-151" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb43-152"><a href="#cb43-152" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(resolution), <span class="at">y =</span> Fstat, </span>
<span id="cb43-153"><a href="#cb43-153" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span> </span>
<span id="cb43-154"><a href="#cb43-154" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb43-155"><a href="#cb43-155" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb43-156"><a href="#cb43-156" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.07</span>) <span class="sc">+</span></span>
<span id="cb43-157"><a href="#cb43-157" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"resolution"</span>)</span>
<span id="cb43-158"><a href="#cb43-158" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb43-159"><a href="#cb43-159" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-160"><a href="#cb43-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-161"><a href="#cb43-161" aria-hidden="true" tabindex="-1"></a><span class="fu">### edgeR</span></span>
<span id="cb43-162"><a href="#cb43-162" aria-hidden="true" tabindex="-1"></a><span class="in">```{r edgeR, fig.height=8, fig.width=15, warning=FALSE}</span></span>
<span id="cb43-163"><a href="#cb43-163" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> res_df[res_df<span class="sc">$</span>edgeR <span class="sc">!=</span> <span class="dv">0</span>, ]</span>
<span id="cb43-164"><a href="#cb43-164" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">na.omit</span>(res_df), <span class="fu">aes</span>(<span class="at">x =</span> edgeR, <span class="at">y =</span> <span class="fu">factor</span>(resolution), </span>
<span id="cb43-165"><a href="#cb43-165" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span></span>
<span id="cb43-166"><a href="#cb43-166" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>() <span class="sc">+</span></span>
<span id="cb43-167"><a href="#cb43-167" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"clustering resolution"</span>) <span class="sc">+</span> </span>
<span id="cb43-168"><a href="#cb43-168" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb43-169"><a href="#cb43-169" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb43-170"><a href="#cb43-170" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"edgeR -log(p_val) vs. clustering resolution"</span>) <span class="sc">+</span></span>
<span id="cb43-171"><a href="#cb43-171" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb43-172"><a href="#cb43-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-173"><a href="#cb43-173" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(resolution), <span class="at">y =</span> edgeR, </span>
<span id="cb43-174"><a href="#cb43-174" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span> </span>
<span id="cb43-175"><a href="#cb43-175" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb43-176"><a href="#cb43-176" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb43-177"><a href="#cb43-177" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.07</span>) <span class="sc">+</span></span>
<span id="cb43-178"><a href="#cb43-178" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"resolution"</span>) <span class="sc">+</span></span>
<span id="cb43-179"><a href="#cb43-179" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"edgeR -log(p_val)"</span>)</span>
<span id="cb43-180"><a href="#cb43-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-181"><a href="#cb43-181" aria-hidden="true" tabindex="-1"></a>p3 <span class="sc">+</span> p4 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb43-182"><a href="#cb43-182" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-183"><a href="#cb43-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-184"><a href="#cb43-184" aria-hidden="true" tabindex="-1"></a><span class="fu">### Fstat_edgeR</span></span>
<span id="cb43-185"><a href="#cb43-185" aria-hidden="true" tabindex="-1"></a><span class="in">```{r Fstat_edgeR, fig.height=8, fig.width=15, warning=FALSE}</span></span>
<span id="cb43-186"><a href="#cb43-186" aria-hidden="true" tabindex="-1"></a>res_df <span class="ot">&lt;-</span> res_df[res_df<span class="sc">$</span>Fstat_edgeR <span class="sc">!=</span> <span class="dv">0</span>, ]</span>
<span id="cb43-187"><a href="#cb43-187" aria-hidden="true" tabindex="-1"></a>p5 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">na.omit</span>(res_df), <span class="fu">aes</span>(<span class="at">x =</span> Fstat_edgeR, <span class="at">y =</span> <span class="fu">factor</span>(resolution), </span>
<span id="cb43-188"><a href="#cb43-188" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span></span>
<span id="cb43-189"><a href="#cb43-189" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density_ridges</span>() <span class="sc">+</span></span>
<span id="cb43-190"><a href="#cb43-190" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"clustering resolution"</span>) <span class="sc">+</span> </span>
<span id="cb43-191"><a href="#cb43-191" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_viridis_d</span>() <span class="sc">+</span></span>
<span id="cb43-192"><a href="#cb43-192" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>() <span class="sc">+</span></span>
<span id="cb43-193"><a href="#cb43-193" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Fstat_edgeR vs. clustering resolution"</span>) <span class="sc">+</span></span>
<span id="cb43-194"><a href="#cb43-194" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span>
<span id="cb43-195"><a href="#cb43-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-196"><a href="#cb43-196" aria-hidden="true" tabindex="-1"></a>p6 <span class="ot">&lt;-</span>  <span class="fu">ggplot</span>(res_df, <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">factor</span>(resolution), <span class="at">y =</span> Fstat_edgeR, </span>
<span id="cb43-197"><a href="#cb43-197" aria-hidden="true" tabindex="-1"></a>  <span class="at">fill =</span> <span class="fu">factor</span>(resolution))) <span class="sc">+</span> </span>
<span id="cb43-198"><a href="#cb43-198" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">alpha =</span> <span class="fl">0.4</span>) <span class="sc">+</span></span>
<span id="cb43-199"><a href="#cb43-199" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb43-200"><a href="#cb43-200" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.07</span>) <span class="sc">+</span></span>
<span id="cb43-201"><a href="#cb43-201" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"resolution"</span>) <span class="sc">+</span></span>
<span id="cb43-202"><a href="#cb43-202" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Fstat_edgeR"</span>)</span>
<span id="cb43-203"><a href="#cb43-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-204"><a href="#cb43-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-205"><a href="#cb43-205" aria-hidden="true" tabindex="-1"></a>p5 <span class="sc">+</span> p6 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb43-206"><a href="#cb43-206" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-207"><a href="#cb43-207" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-208"><a href="#cb43-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-209"><a href="#cb43-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-212"><a href="#cb43-212" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-213"><a href="#cb43-213" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Initial clustering</span></span>
<span id="cb43-214"><a href="#cb43-214" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb43-215"><a href="#cb43-215" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">buildSNNGraph</span>(sce, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb43-216"><a href="#cb43-216" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_hi <span class="ot">&lt;-</span> <span class="fu">cluster_louvain</span>(g, <span class="at">resolution =</span> <span class="dv">5</span>)<span class="sc">$</span>membership</span>
<span id="cb43-217"><a href="#cb43-217" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_lo <span class="ot">&lt;-</span> <span class="fu">cluster_louvain</span>(g, <span class="at">resolution =</span> <span class="fl">0.6</span>)<span class="sc">$</span>membership</span>
<span id="cb43-218"><a href="#cb43-218" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_hi <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce<span class="sc">$</span>cluster_hi, <span class="fu">sort</span>(<span class="fu">unique</span>(sce<span class="sc">$</span>cluster_hi)))</span>
<span id="cb43-219"><a href="#cb43-219" aria-hidden="true" tabindex="-1"></a>sce<span class="sc">$</span>cluster_lo <span class="ot">&lt;-</span> <span class="fu">factor</span>(sce<span class="sc">$</span>cluster_lo, <span class="fu">sort</span>(<span class="fu">unique</span>(sce<span class="sc">$</span>cluster_lo)))</span>
<span id="cb43-220"><a href="#cb43-220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-221"><a href="#cb43-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-222"><a href="#cb43-222" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb43-223"><a href="#cb43-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-224"><a href="#cb43-224" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA - hvg, high resolution</span></span>
<span id="cb43-227"><a href="#cb43-227" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-228"><a href="#cb43-228" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: PCA high</span></span>
<span id="cb43-229"><a href="#cb43-229" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"cluster_hi"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span>
<span id="cb43-230"><a href="#cb43-230" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-231"><a href="#cb43-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-232"><a href="#cb43-232" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA - hvg, low resolution</span></span>
<span id="cb43-235"><a href="#cb43-235" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-236"><a href="#cb43-236" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: PCA low</span></span>
<span id="cb43-237"><a href="#cb43-237" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(sce, <span class="at">colour_by =</span> <span class="st">"cluster_lo"</span>) <span class="sc">+</span> <span class="fu">coord_equal</span>()</span>
<span id="cb43-238"><a href="#cb43-238" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-239"><a href="#cb43-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-240"><a href="#cb43-240" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-241"><a href="#cb43-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-242"><a href="#cb43-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-243"><a href="#cb43-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-244"><a href="#cb43-244" aria-hidden="true" tabindex="-1"></a><span class="fu">## Feature scoring</span></span>
<span id="cb43-245"><a href="#cb43-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-246"><a href="#cb43-246" aria-hidden="true" tabindex="-1"></a>Here, we calculated two type scores (F-statistics and entropy), one state scores (p values of DS analysis by edgeR) and random score drawed from uniform distribution ranging from 0 to 1. The following figure illustrates the relationship between different scores colored by HVG. We also added two published feature selection tools into comparison: *FEAST* which is also a clustering-based filter method scoring the genes by F-statistics, and *DUBStepR* that is based on the idea of cell-type-specific markers should be well correlated with each other.</span>
<span id="cb43-247"><a href="#cb43-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-248"><a href="#cb43-248" aria-hidden="true" tabindex="-1"></a><span class="in">```{r warning=FALSE, message=FALSE, results=FALSE, fig.height=6, fig.width=9}</span></span>
<span id="cb43-249"><a href="#cb43-249" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Score features</span></span>
<span id="cb43-250"><a href="#cb43-250" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat <span class="ot">&lt;-</span> <span class="fu">.Fstat</span>(sce)</span>
<span id="cb43-251"><a href="#cb43-251" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR <span class="ot">&lt;-</span> <span class="fu">.edgeR</span>(sce)</span>
<span id="cb43-252"><a href="#cb43-252" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>random <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="fu">nrow</span>(sce))</span>
<span id="cb43-253"><a href="#cb43-253" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR <span class="ot">&lt;-</span> <span class="fu">rank</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat) <span class="sc">-</span> <span class="fu">rank</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR)</span>
<span id="cb43-254"><a href="#cb43-254" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST <span class="ot">&lt;-</span> <span class="fu">.FEAST</span>(sce)</span>
<span id="cb43-255"><a href="#cb43-255" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR <span class="ot">&lt;-</span> <span class="fu">.DUBStepR</span>(sce)</span>
<span id="cb43-256"><a href="#cb43-256" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-257"><a href="#cb43-257" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">rowData</span>(sce)) <span class="sc">%&gt;%</span> </span>
<span id="cb43-258"><a href="#cb43-258" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">Fstat =</span> <span class="fu">log10</span>(Fstat),</span>
<span id="cb43-259"><a href="#cb43-259" aria-hidden="true" tabindex="-1"></a>        <span class="at">edgeR =</span> <span class="fu">log10</span>(edgeR))</span>
<span id="cb43-260"><a href="#cb43-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-261"><a href="#cb43-261" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(df, <span class="at">columns =</span> <span class="dv">6</span><span class="sc">:</span><span class="dv">11</span>, </span>
<span id="cb43-262"><a href="#cb43-262" aria-hidden="true" tabindex="-1"></a>    ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">colour =</span> hvg, <span class="at">alpha =</span> <span class="fl">0.4</span>),</span>
<span id="cb43-263"><a href="#cb43-263" aria-hidden="true" tabindex="-1"></a>    <span class="at">upper =</span> <span class="st">"blank"</span>,</span>
<span id="cb43-264"><a href="#cb43-264" aria-hidden="true" tabindex="-1"></a>    <span class="at">legend =</span> <span class="dv">1</span>)</span>
<span id="cb43-265"><a href="#cb43-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-266"><a href="#cb43-266" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-267"><a href="#cb43-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-268"><a href="#cb43-268" aria-hidden="true" tabindex="-1"></a><span class="fu">### Further exploration of different scores</span></span>
<span id="cb43-269"><a href="#cb43-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-270"><a href="#cb43-270" aria-hidden="true" tabindex="-1"></a>We utilized the <span class="in">`isOutlier`</span> function to identify genes whose type/state scores were outliers, using the median absolute deviation (MAD) as the criterion. For highly variable genes (HVG), we selected those with a biological variation greater than 0. Furthermore, **DUBStepR** has already furnished an optimal set of features. An ideal marker is expected to exhibit a high Fstat score but a low edgeR score, signifying a high type score yet a low state score. To ensure each method selected a sufficient number of genes, we set the <span class="in">`nmads`</span> parameter to 1.</span>
<span id="cb43-271"><a href="#cb43-271" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-274"><a href="#cb43-274" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-275"><a href="#cb43-275" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="dv">1</span>)</span>
<span id="cb43-276"><a href="#cb43-276" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="dv">1</span>)</span>
<span id="cb43-277"><a href="#cb43-277" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="fl">0.1</span>)</span>
<span id="cb43-278"><a href="#cb43-278" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg_sel <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>bio <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb43-279"><a href="#cb43-279" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST_sel <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(<span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST, <span class="at">type =</span> <span class="st">"higher"</span>, <span class="at">nmads =</span> <span class="dv">1</span>)</span>
<span id="cb43-280"><a href="#cb43-280" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR <span class="sc">&gt;</span> <span class="dv">0</span></span>
<span id="cb43-281"><a href="#cb43-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-282"><a href="#cb43-282" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Fstat =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_sel,<span class="st">"gene_id"</span>], </span>
<span id="cb43-283"><a href="#cb43-283" aria-hidden="true" tabindex="-1"></a>  <span class="at">edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-284"><a href="#cb43-284" aria-hidden="true" tabindex="-1"></a>  <span class="at">Fstat_edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-285"><a href="#cb43-285" aria-hidden="true" tabindex="-1"></a>  <span class="at">hvg =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-286"><a href="#cb43-286" aria-hidden="true" tabindex="-1"></a>  <span class="at">FEAST =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-287"><a href="#cb43-287" aria-hidden="true" tabindex="-1"></a>  <span class="at">DUBStepR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel,<span class="st">"gene_id"</span>])</span>
<span id="cb43-288"><a href="#cb43-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-289"><a href="#cb43-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-290"><a href="#cb43-290" aria-hidden="true" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(lst),</span>
<span id="cb43-291"><a href="#cb43-291" aria-hidden="true" tabindex="-1"></a>        <span class="at">order.by =</span> <span class="st">"freq"</span>,</span>
<span id="cb43-292"><a href="#cb43-292" aria-hidden="true" tabindex="-1"></a>        <span class="at">nsets =</span> <span class="fu">length</span>(lst),</span>
<span id="cb43-293"><a href="#cb43-293" aria-hidden="true" tabindex="-1"></a>        <span class="at">set_size.show =</span> <span class="cn">TRUE</span>,</span>
<span id="cb43-294"><a href="#cb43-294" aria-hidden="true" tabindex="-1"></a>        <span class="at">scale.sets =</span> <span class="st">"identity"</span>)</span>
<span id="cb43-295"><a href="#cb43-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-296"><a href="#cb43-296" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-297"><a href="#cb43-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-298"><a href="#cb43-298" aria-hidden="true" tabindex="-1"></a>Here, we extract genes with rank difference between Fstat type score and edgeR state score larger than 0, indicating a gene has relatively larger type score than state score. It is observed that even genes with every high F-statistics values but high edgeR value might still be excluded by *Fstat_edgeR* selection. </span>
<span id="cb43-299"><a href="#cb43-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-302"><a href="#cb43-302" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-303"><a href="#cb43-303" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb43-304"><a href="#cb43-304" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="fu">data.frame</span>(sel), <span class="fu">aes</span>(<span class="at">x =</span> <span class="fu">log10</span>(Fstat), <span class="at">y =</span> edgeR, <span class="at">col =</span> Fstat_edgeR)) <span class="sc">+</span> </span>
<span id="cb43-305"><a href="#cb43-305" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb43-306"><a href="#cb43-306" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradient2</span>() <span class="sc">+</span></span>
<span id="cb43-307"><a href="#cb43-307" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"-log(p_val) in edgeR DS"</span>)</span>
<span id="cb43-308"><a href="#cb43-308" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-309"><a href="#cb43-309" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-310"><a href="#cb43-310" aria-hidden="true" tabindex="-1"></a><span class="fu">### Determine the optimal size of features</span></span>
<span id="cb43-311"><a href="#cb43-311" aria-hidden="true" tabindex="-1"></a>To determine the optimal size of selected features, we tried different number of genes and evaluate its performance using ARI, mean silhouette score and neighbour purity between clusters. There are two dashed lines: the black dashed line indicates the evaluation value when no feature selection, i.e. using all features; the red dashed line shows the results by **DUBStepR** because it provides an optimal set of features and thus no need to vary the feature size. </span>
<span id="cb43-312"><a href="#cb43-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-313"><a href="#cb43-313" aria-hidden="true" tabindex="-1"></a><span class="in">```{r warning=FALSE, results='hide'}</span></span>
<span id="cb43-314"><a href="#cb43-314" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Evaluating number of selected features</span></span>
<span id="cb43-315"><a href="#cb43-315" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR"</span>, <span class="st">"random"</span>, <span class="st">"bio"</span>, <span class="st">"Fstat"</span>, <span class="st">"FEAST"</span>)</span>
<span id="cb43-316"><a href="#cb43-316" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fl">0.1</span>,<span class="dv">1</span>,<span class="fl">0.05</span>)</span>
<span id="cb43-317"><a href="#cb43-317" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-318"><a href="#cb43-318" aria-hidden="true" tabindex="-1"></a>params <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">n =</span> n, <span class="at">sel =</span> sel)</span>
<span id="cb43-319"><a href="#cb43-319" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-320"><a href="#cb43-320" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop over each row of params and compute performance</span></span>
<span id="cb43-321"><a href="#cb43-321" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(params), \(i) {</span>
<span id="cb43-322"><a href="#cb43-322" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb43-323"><a href="#cb43-323" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> params<span class="sc">$</span>n[i]</span>
<span id="cb43-324"><a href="#cb43-324" aria-hidden="true" tabindex="-1"></a>  s <span class="ot">&lt;-</span> params<span class="sc">$</span>sel[i]</span>
<span id="cb43-325"><a href="#cb43-325" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowData</span>(sce)[, s], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="fu">seq_len</span>(<span class="fu">round</span>(<span class="fu">nrow</span>(sce)<span class="sc">*</span>x))]</span>
<span id="cb43-326"><a href="#cb43-326" aria-hidden="true" tabindex="-1"></a>  sel_val <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">nrow</span>(sce))</span>
<span id="cb43-327"><a href="#cb43-327" aria-hidden="true" tabindex="-1"></a>  sel_val[idx] <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb43-328"><a href="#cb43-328" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> sel_val)</span>
<span id="cb43-329"><a href="#cb43-329" aria-hidden="true" tabindex="-1"></a>  sce<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb43-330"><a href="#cb43-330" aria-hidden="true" tabindex="-1"></a>  ari <span class="ot">&lt;-</span> <span class="fu">ARI</span>(sce<span class="sc">$</span>cluster_re, sce<span class="sc">$</span>cluster_id)</span>
<span id="cb43-331"><a href="#cb43-331" aria-hidden="true" tabindex="-1"></a>  sil <span class="ot">&lt;-</span> <span class="fu">.sil_k</span>(sce)</span>
<span id="cb43-332"><a href="#cb43-332" aria-hidden="true" tabindex="-1"></a>  pur <span class="ot">&lt;-</span> <span class="fu">.pur_k</span>(sce)</span>
<span id="cb43-333"><a href="#cb43-333" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>(<span class="at">ARI =</span> ari, <span class="at">sil =</span> sil, <span class="at">pur =</span> pur, <span class="at">sel =</span> s, <span class="at">n =</span> x)</span>
<span id="cb43-334"><a href="#cb43-334" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-335"><a href="#cb43-335" aria-hidden="true" tabindex="-1"></a>sta <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">do.call</span>(rbind, res))</span>
<span id="cb43-336"><a href="#cb43-336" aria-hidden="true" tabindex="-1"></a><span class="co">#saveRDS(sta, "../data/case/sta.rds")</span></span>
<span id="cb43-337"><a href="#cb43-337" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-338"><a href="#cb43-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-339"><a href="#cb43-339" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=8}</span></span>
<span id="cb43-340"><a href="#cb43-340" aria-hidden="true" tabindex="-1"></a><span class="co">#sta &lt;- readRDS("/Volumes/jiayiwang/type-state/data/case/sta.rds")</span></span>
<span id="cb43-341"><a href="#cb43-341" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">ARI =</span> sta<span class="sc">$</span>ARI,</span>
<span id="cb43-342"><a href="#cb43-342" aria-hidden="true" tabindex="-1"></a>    <span class="at">precentage =</span> n,</span>
<span id="cb43-343"><a href="#cb43-343" aria-hidden="true" tabindex="-1"></a>    <span class="at">silhouette =</span> sta<span class="sc">$</span>sil,</span>
<span id="cb43-344"><a href="#cb43-344" aria-hidden="true" tabindex="-1"></a>    <span class="at">purity =</span> sta<span class="sc">$</span>pur,</span>
<span id="cb43-345"><a href="#cb43-345" aria-hidden="true" tabindex="-1"></a>    <span class="at">number =</span> <span class="fu">nrow</span>(sce)<span class="sc">*</span>n,</span>
<span id="cb43-346"><a href="#cb43-346" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> sta<span class="sc">$</span>sel) <span class="sc">%&gt;%</span> </span>
<span id="cb43-347"><a href="#cb43-347" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">paste0</span>(number, <span class="st">", "</span>, <span class="fu">round</span>(ARI, <span class="at">digits =</span> <span class="dv">2</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb43-348"><a href="#cb43-348" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">sel =</span> <span class="fu">case_when</span>(sel <span class="sc">==</span> <span class="st">"bio"</span> <span class="sc">~</span> <span class="st">"hvg"</span>,</span>
<span id="cb43-349"><a href="#cb43-349" aria-hidden="true" tabindex="-1"></a>      <span class="cn">TRUE</span> <span class="sc">~</span> sel))</span>
<span id="cb43-350"><a href="#cb43-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-351"><a href="#cb43-351" aria-hidden="true" tabindex="-1"></a><span class="co"># baseline</span></span>
<span id="cb43-352"><a href="#cb43-352" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">2023</span>)</span>
<span id="cb43-353"><a href="#cb43-353" aria-hidden="true" tabindex="-1"></a>all <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">TRUE</span>, <span class="fu">nrow</span>(sce))</span>
<span id="cb43-354"><a href="#cb43-354" aria-hidden="true" tabindex="-1"></a>sce1 <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> all)</span>
<span id="cb43-355"><a href="#cb43-355" aria-hidden="true" tabindex="-1"></a>sce1<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce1, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb43-356"><a href="#cb43-356" aria-hidden="true" tabindex="-1"></a>base_ari <span class="ot">&lt;-</span> <span class="fu">ARI</span>(sce1<span class="sc">$</span>cluster_re, sce1<span class="sc">$</span>cluster_id)</span>
<span id="cb43-357"><a href="#cb43-357" aria-hidden="true" tabindex="-1"></a>base_sil <span class="ot">&lt;-</span> <span class="fu">.sil_k</span>(sce1)</span>
<span id="cb43-358"><a href="#cb43-358" aria-hidden="true" tabindex="-1"></a>base_pur <span class="ot">&lt;-</span> <span class="fu">.pur_k</span>(sce1)</span>
<span id="cb43-359"><a href="#cb43-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-360"><a href="#cb43-360" aria-hidden="true" tabindex="-1"></a><span class="co"># DUBStepR</span></span>
<span id="cb43-361"><a href="#cb43-361" aria-hidden="true" tabindex="-1"></a>sce2 <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> <span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel)</span>
<span id="cb43-362"><a href="#cb43-362" aria-hidden="true" tabindex="-1"></a>sce2<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce2, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb43-363"><a href="#cb43-363" aria-hidden="true" tabindex="-1"></a>dub_ari <span class="ot">&lt;-</span> <span class="fu">ARI</span>(sce2<span class="sc">$</span>cluster_re, sce2<span class="sc">$</span>cluster_id)</span>
<span id="cb43-364"><a href="#cb43-364" aria-hidden="true" tabindex="-1"></a>dub_sil <span class="ot">&lt;-</span> <span class="fu">.sil_k</span>(sce2)</span>
<span id="cb43-365"><a href="#cb43-365" aria-hidden="true" tabindex="-1"></a>dub_pur <span class="ot">&lt;-</span> <span class="fu">.pur_k</span>(sce2)</span>
<span id="cb43-366"><a href="#cb43-366" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-367"><a href="#cb43-367" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-368"><a href="#cb43-368" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb43-369"><a href="#cb43-369" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-370"><a href="#cb43-370" aria-hidden="true" tabindex="-1"></a><span class="fu">### ARI</span></span>
<span id="cb43-371"><a href="#cb43-371" aria-hidden="true" tabindex="-1"></a><span class="in">```{r , fig.width=8}</span></span>
<span id="cb43-372"><a href="#cb43-372" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> precentage, <span class="at">y =</span> ARI, <span class="at">col =</span> sel)) <span class="sc">+</span> </span>
<span id="cb43-373"><a href="#cb43-373" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb43-374"><a href="#cb43-374" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb43-375"><a href="#cb43-375" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb43-376"><a href="#cb43-376" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> base_ari, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb43-377"><a href="#cb43-377" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-378"><a href="#cb43-378" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> dub_ari, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb43-379"><a href="#cb43-379" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-380"><a href="#cb43-380" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">0.1</span>, <span class="at">y =</span> base_ari <span class="sc">+</span> <span class="fl">0.02</span>, </span>
<span id="cb43-381"><a href="#cb43-381" aria-hidden="true" tabindex="-1"></a>       <span class="at">label =</span> <span class="st">"no feature selection"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb43-382"><a href="#cb43-382" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"ARI vs. precentage of the selected features"</span>) <span class="sc">+</span></span>
<span id="cb43-383"><a href="#cb43-383" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb43-384"><a href="#cb43-384" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-385"><a href="#cb43-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-386"><a href="#cb43-386" aria-hidden="true" tabindex="-1"></a><span class="fu">### Silhouette Score</span></span>
<span id="cb43-387"><a href="#cb43-387" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, fig.width=8}</span></span>
<span id="cb43-388"><a href="#cb43-388" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> precentage, <span class="at">y =</span> silhouette, <span class="at">col =</span> sel)) <span class="sc">+</span> </span>
<span id="cb43-389"><a href="#cb43-389" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb43-390"><a href="#cb43-390" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb43-391"><a href="#cb43-391" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> base_sil, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb43-392"><a href="#cb43-392" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-393"><a href="#cb43-393" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> dub_sil, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb43-394"><a href="#cb43-394" aria-hidden="true" tabindex="-1"></a>                    <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-395"><a href="#cb43-395" aria-hidden="true" tabindex="-1"></a>     <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">0.1</span>, <span class="at">y =</span> base_sil <span class="sc">+</span> <span class="fl">0.01</span>, </span>
<span id="cb43-396"><a href="#cb43-396" aria-hidden="true" tabindex="-1"></a>       <span class="at">label =</span> <span class="st">"no feature selection"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb43-397"><a href="#cb43-397" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb43-398"><a href="#cb43-398" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"silhouette vs. precentage of the selected features"</span>) <span class="sc">+</span></span>
<span id="cb43-399"><a href="#cb43-399" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb43-400"><a href="#cb43-400" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-401"><a href="#cb43-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-402"><a href="#cb43-402" aria-hidden="true" tabindex="-1"></a><span class="fu">### Neighbourhood purity</span></span>
<span id="cb43-403"><a href="#cb43-403" aria-hidden="true" tabindex="-1"></a><span class="in">```{r , fig.width=8}</span></span>
<span id="cb43-404"><a href="#cb43-404" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> precentage, <span class="at">y =</span> purity, <span class="at">col =</span> sel)) <span class="sc">+</span> </span>
<span id="cb43-405"><a href="#cb43-405" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_line</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb43-406"><a href="#cb43-406" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb43-407"><a href="#cb43-407" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> base_pur, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb43-408"><a href="#cb43-408" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-409"><a href="#cb43-409" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> dub_pur, <span class="at">linetype =</span> <span class="st">"dashed"</span>, </span>
<span id="cb43-410"><a href="#cb43-410" aria-hidden="true" tabindex="-1"></a>                      <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb43-411"><a href="#cb43-411" aria-hidden="true" tabindex="-1"></a>     <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">0.1</span>, <span class="at">y =</span> base_pur <span class="sc">-</span> <span class="fl">0.03</span>, </span>
<span id="cb43-412"><a href="#cb43-412" aria-hidden="true" tabindex="-1"></a>       <span class="at">label =</span> <span class="st">"no feature selection"</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb43-413"><a href="#cb43-413" aria-hidden="true" tabindex="-1"></a>    <span class="fu">xlim</span>(<span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb43-414"><a href="#cb43-414" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"NeighbourPurity vs. precentage of the selected features"</span>) <span class="sc">+</span></span>
<span id="cb43-415"><a href="#cb43-415" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>)</span>
<span id="cb43-416"><a href="#cb43-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-417"><a href="#cb43-417" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-418"><a href="#cb43-418" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-419"><a href="#cb43-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-420"><a href="#cb43-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-421"><a href="#cb43-421" aria-hidden="true" tabindex="-1"></a><span class="fu">## Feature selection and reprosessing </span></span>
<span id="cb43-422"><a href="#cb43-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-423"><a href="#cb43-423" aria-hidden="true" tabindex="-1"></a><span class="fu">### Feature selection step</span></span>
<span id="cb43-424"><a href="#cb43-424" aria-hidden="true" tabindex="-1"></a>The above evaluation statistics results suggested that different selection methods have their own specific optimal size. Now we simply selected top 40% genes based on their scores (except for hvg and **DUBStepR**), which requires further adaption of the feature sizes. </span>
<span id="cb43-425"><a href="#cb43-425" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-428"><a href="#cb43-428" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb43-429"><a href="#cb43-429" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Feature selection and reprocess</span></span>
<span id="cb43-430"><a href="#cb43-430" aria-hidden="true" tabindex="-1"></a>fs <span class="ot">&lt;-</span> \(sce, n, sel, res) {</span>
<span id="cb43-431"><a href="#cb43-431" aria-hidden="true" tabindex="-1"></a>    idx <span class="ot">&lt;-</span> <span class="fu">order</span>(<span class="fu">rowData</span>(sce)[,sel], </span>
<span id="cb43-432"><a href="#cb43-432" aria-hidden="true" tabindex="-1"></a>      <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="fu">seq_len</span>(<span class="fu">round</span>(<span class="fu">nrow</span>(sce)<span class="sc">*</span>n))]</span>
<span id="cb43-433"><a href="#cb43-433" aria-hidden="true" tabindex="-1"></a>    sel_val <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">FALSE</span>, <span class="fu">nrow</span>(sce))</span>
<span id="cb43-434"><a href="#cb43-434" aria-hidden="true" tabindex="-1"></a>    sel_val[idx] <span class="ot">&lt;-</span> <span class="cn">TRUE</span></span>
<span id="cb43-435"><a href="#cb43-435" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rowData</span>(sce)[, res] <span class="ot">&lt;-</span> sel_val</span>
<span id="cb43-436"><a href="#cb43-436" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(sce)</span>
<span id="cb43-437"><a href="#cb43-437" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-438"><a href="#cb43-438" aria-hidden="true" tabindex="-1"></a>sco <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR"</span>, <span class="st">"Fstat"</span>, <span class="st">"random"</span>, <span class="st">"FEAST"</span>)</span>
<span id="cb43-439"><a href="#cb43-439" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (s <span class="cf">in</span> sel) {</span>
<span id="cb43-440"><a href="#cb43-440" aria-hidden="true" tabindex="-1"></a>  r <span class="ot">&lt;-</span> <span class="fu">paste0</span>(s,<span class="st">"_sel"</span>)</span>
<span id="cb43-441"><a href="#cb43-441" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">fs</span>(sce, <span class="at">n =</span> <span class="fl">0.4</span>, <span class="at">sel =</span> s, <span class="at">res =</span> r)</span>
<span id="cb43-442"><a href="#cb43-442" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb43-443"><a href="#cb43-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-444"><a href="#cb43-444" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR_sel"</span>, <span class="st">"Fstat_sel"</span>, <span class="st">"random_sel"</span>, </span>
<span id="cb43-445"><a href="#cb43-445" aria-hidden="true" tabindex="-1"></a>    <span class="st">"hvg"</span>, <span class="st">"DUBStepR_sel"</span>, <span class="st">"FEAST_sel"</span>)</span>
<span id="cb43-446"><a href="#cb43-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-447"><a href="#cb43-447" aria-hidden="true" tabindex="-1"></a>sces <span class="ot">&lt;-</span> <span class="fu">lapply</span>(sel, \(s) {</span>
<span id="cb43-448"><a href="#cb43-448" aria-hidden="true" tabindex="-1"></a>  sce <span class="ot">&lt;-</span> <span class="fu">runPCA</span>(sce, <span class="at">subset_row =</span> <span class="fu">rowData</span>(sce)[,s])</span>
<span id="cb43-449"><a href="#cb43-449" aria-hidden="true" tabindex="-1"></a>  sce<span class="sc">$</span>cluster_re <span class="ot">&lt;-</span> <span class="fu">clusterCells</span>(sce, <span class="at">use.dimred =</span> <span class="st">"PCA"</span>)</span>
<span id="cb43-450"><a href="#cb43-450" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(sce)</span>
<span id="cb43-451"><a href="#cb43-451" aria-hidden="true" tabindex="-1"></a>}) </span>
<span id="cb43-452"><a href="#cb43-452" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(sces) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Fstat_edgeR"</span>, <span class="st">"Fstat"</span>, <span class="st">"random"</span>, <span class="st">"hvg"</span>, <span class="st">"DUBStepR"</span>, <span class="st">"FEAST"</span>)</span>
<span id="cb43-453"><a href="#cb43-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-454"><a href="#cb43-454" aria-hidden="true" tabindex="-1"></a>lst <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">Fstat =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_sel,<span class="st">"gene_id"</span>], </span>
<span id="cb43-455"><a href="#cb43-455" aria-hidden="true" tabindex="-1"></a>  <span class="at">edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-456"><a href="#cb43-456" aria-hidden="true" tabindex="-1"></a>  <span class="at">Fstat_edgeR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>Fstat_edgeR_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-457"><a href="#cb43-457" aria-hidden="true" tabindex="-1"></a>  <span class="at">hvg =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>hvg_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-458"><a href="#cb43-458" aria-hidden="true" tabindex="-1"></a>  <span class="at">FEAST =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>FEAST_sel,<span class="st">"gene_id"</span>],</span>
<span id="cb43-459"><a href="#cb43-459" aria-hidden="true" tabindex="-1"></a>  <span class="at">DUBStepR =</span> <span class="fu">rowData</span>(sce)[<span class="fu">rowData</span>(sce)<span class="sc">$</span>DUBStepR_sel,<span class="st">"gene_id"</span>])</span>
<span id="cb43-460"><a href="#cb43-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-461"><a href="#cb43-461" aria-hidden="true" tabindex="-1"></a><span class="fu">upset</span>(<span class="fu">fromList</span>(lst),</span>
<span id="cb43-462"><a href="#cb43-462" aria-hidden="true" tabindex="-1"></a>        <span class="at">order.by =</span> <span class="st">"freq"</span>,</span>
<span id="cb43-463"><a href="#cb43-463" aria-hidden="true" tabindex="-1"></a>        <span class="at">nsets =</span> <span class="fu">length</span>(lst),</span>
<span id="cb43-464"><a href="#cb43-464" aria-hidden="true" tabindex="-1"></a>        <span class="at">set_size.show =</span> <span class="cn">TRUE</span>,</span>
<span id="cb43-465"><a href="#cb43-465" aria-hidden="true" tabindex="-1"></a>        <span class="at">scale.sets =</span> <span class="st">"identity"</span>)</span>
<span id="cb43-466"><a href="#cb43-466" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-467"><a href="#cb43-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-468"><a href="#cb43-468" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA and UMAP on reprocessed data</span></span>
<span id="cb43-469"><a href="#cb43-469" aria-hidden="true" tabindex="-1"></a>PCA plots on selected features indicated that all other methods are affected by the strong state effect in the data, with two conditions clearly separated. However, our Fstat_edgeR gives two blobs with two conditions are well mixed. </span>
<span id="cb43-470"><a href="#cb43-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-471"><a href="#cb43-471" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb43-472"><a href="#cb43-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-473"><a href="#cb43-473" aria-hidden="true" tabindex="-1"></a><span class="fu">#### new PCA, cluster_re</span></span>
<span id="cb43-474"><a href="#cb43-474" aria-hidden="true" tabindex="-1"></a><span class="in">```{r pk, warning=FALSE, fig.height=10, fig.width=8}</span></span>
<span id="cb43-475"><a href="#cb43-475" aria-hidden="true" tabindex="-1"></a>pk <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) <span class="fu">plotPCA</span>(sces[[x]], <span class="at">colour_by =</span> <span class="st">"cluster_re"</span>) <span class="sc">+</span> </span>
<span id="cb43-476"><a href="#cb43-476" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(x)) <span class="sc">|&gt;</span> </span>
<span id="cb43-477"><a href="#cb43-477" aria-hidden="true" tabindex="-1"></a>    <span class="fu">wrap_plots</span>(<span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span>
<span id="cb43-478"><a href="#cb43-478" aria-hidden="true" tabindex="-1"></a>pk</span>
<span id="cb43-479"><a href="#cb43-479" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-480"><a href="#cb43-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-481"><a href="#cb43-481" aria-hidden="true" tabindex="-1"></a><span class="fu">#### new PCA, condition</span></span>
<span id="cb43-482"><a href="#cb43-482" aria-hidden="true" tabindex="-1"></a><span class="in">```{r pg, warning=FALSE, fig.height=10, fig.width=8}</span></span>
<span id="cb43-483"><a href="#cb43-483" aria-hidden="true" tabindex="-1"></a>pg <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) <span class="fu">plotPCA</span>(sces[[x]], <span class="at">colour_by =</span> <span class="st">"group_id"</span>) <span class="sc">+</span> </span>
<span id="cb43-484"><a href="#cb43-484" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(x)) <span class="sc">|&gt;</span> </span>
<span id="cb43-485"><a href="#cb43-485" aria-hidden="true" tabindex="-1"></a>    <span class="fu">wrap_plots</span>(<span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span>
<span id="cb43-486"><a href="#cb43-486" aria-hidden="true" tabindex="-1"></a>pg</span>
<span id="cb43-487"><a href="#cb43-487" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-488"><a href="#cb43-488" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-489"><a href="#cb43-489" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-490"><a href="#cb43-490" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-491"><a href="#cb43-491" aria-hidden="true" tabindex="-1"></a><span class="fu">### Abundance plot </span></span>
<span id="cb43-492"><a href="#cb43-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-493"><a href="#cb43-493" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset}</span>
<span id="cb43-494"><a href="#cb43-494" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-495"><a href="#cb43-495" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Proportion of cell types in each cluster_re</span></span>
<span id="cb43-496"><a href="#cb43-496" aria-hidden="true" tabindex="-1"></a><span class="in">```{r abundance plot1, fig.height=10, fig.width=12}</span></span>
<span id="cb43-497"><a href="#cb43-497" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: Abundance</span></span>
<span id="cb43-498"><a href="#cb43-498" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x){</span>
<span id="cb43-499"><a href="#cb43-499" aria-hidden="true" tabindex="-1"></a>    cd <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(sces[[x]]))</span>
<span id="cb43-500"><a href="#cb43-500" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(cd) <span class="sc">+</span></span>
<span id="cb43-501"><a href="#cb43-501" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">fill =</span> cluster_id) <span class="sc">+</span></span>
<span id="cb43-502"><a href="#cb43-502" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb43-503"><a href="#cb43-503" aria-hidden="true" tabindex="-1"></a>        <span class="fu">xlab</span>(<span class="st">"Cluster after reprocessed"</span>) <span class="sc">+</span></span>
<span id="cb43-504"><a href="#cb43-504" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>) <span class="sc">+</span> </span>
<span id="cb43-505"><a href="#cb43-505" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggtitle</span>(x)</span>
<span id="cb43-506"><a href="#cb43-506" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">wrap_plots</span>(<span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span>
<span id="cb43-507"><a href="#cb43-507" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb43-508"><a href="#cb43-508" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-509"><a href="#cb43-509" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-510"><a href="#cb43-510" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Proportion of cluster_re in each cell type</span></span>
<span id="cb43-511"><a href="#cb43-511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r abundance plot2, fig.height=10, fig.width=10, warning=FALSE}</span></span>
<span id="cb43-512"><a href="#cb43-512" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x){</span>
<span id="cb43-513"><a href="#cb43-513" aria-hidden="true" tabindex="-1"></a>    cd <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(sces[[x]]))</span>
<span id="cb43-514"><a href="#cb43-514" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(cd) <span class="sc">+</span></span>
<span id="cb43-515"><a href="#cb43-515" aria-hidden="true" tabindex="-1"></a>        <span class="fu">aes</span>(<span class="at">x =</span> cluster_id, <span class="at">fill =</span> cluster_re) <span class="sc">+</span></span>
<span id="cb43-516"><a href="#cb43-516" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"fill"</span>) <span class="sc">+</span></span>
<span id="cb43-517"><a href="#cb43-517" aria-hidden="true" tabindex="-1"></a>        <span class="fu">xlab</span>(<span class="st">"Cell Type"</span>) <span class="sc">+</span></span>
<span id="cb43-518"><a href="#cb43-518" aria-hidden="true" tabindex="-1"></a>        <span class="fu">scale_fill_brewer</span>(<span class="at">palette =</span> <span class="st">"Paired"</span>) <span class="sc">+</span> </span>
<span id="cb43-519"><a href="#cb43-519" aria-hidden="true" tabindex="-1"></a>        <span class="fu">ggtitle</span>(x) <span class="sc">+</span> </span>
<span id="cb43-520"><a href="#cb43-520" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">vjust =</span> <span class="dv">1</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span>
<span id="cb43-521"><a href="#cb43-521" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">wrap_plots</span>(<span class="at">nrow =</span> <span class="dv">2</span>) <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span>
<span id="cb43-522"><a href="#cb43-522" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb43-523"><a href="#cb43-523" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-524"><a href="#cb43-524" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-525"><a href="#cb43-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-526"><a href="#cb43-526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-527"><a href="#cb43-527" aria-hidden="true" tabindex="-1"></a><span class="fu">## Differential analysis</span></span>
<span id="cb43-528"><a href="#cb43-528" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-529"><a href="#cb43-529" aria-hidden="true" tabindex="-1"></a><span class="fu">### DA</span></span>
<span id="cb43-530"><a href="#cb43-530" aria-hidden="true" tabindex="-1"></a><span class="in">```{r DA, message=FALSE, warning=FALSE}</span></span>
<span id="cb43-531"><a href="#cb43-531" aria-hidden="true" tabindex="-1"></a><span class="co"># edgeR</span></span>
<span id="cb43-532"><a href="#cb43-532" aria-hidden="true" tabindex="-1"></a>dal <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) {</span>
<span id="cb43-533"><a href="#cb43-533" aria-hidden="true" tabindex="-1"></a>  e <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DA_edgeR</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, <span class="st">"cluster_re"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb43-534"><a href="#cb43-534" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DA_edgeR"</span>, <span class="at">DADS =</span> <span class="st">"DA"</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>) </span>
<span id="cb43-535"><a href="#cb43-535" aria-hidden="true" tabindex="-1"></a>  l <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DA_limma</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, <span class="st">"cluster_re"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb43-536"><a href="#cb43-536" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DA_limma"</span>, <span class="at">DADS =</span> <span class="st">"DA"</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>)</span>
<span id="cb43-537"><a href="#cb43-537" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DA_milo</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, <span class="st">"Nhood"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb43-538"><a href="#cb43-538" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DA_milo"</span>, <span class="at">DADS =</span> <span class="st">"DA"</span>, <span class="at">clustering =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb43-539"><a href="#cb43-539" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="st">"cluster_re"</span> <span class="ot">=</span> <span class="st">"Nhood"</span>)</span>
<span id="cb43-540"><a href="#cb43-540" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(rbind, <span class="fu">list</span>(e,l,m))</span>
<span id="cb43-541"><a href="#cb43-541" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-542"><a href="#cb43-542" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-543"><a href="#cb43-543" aria-hidden="true" tabindex="-1"></a>da <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, dal)</span>
<span id="cb43-544"><a href="#cb43-544" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-545"><a href="#cb43-545" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-546"><a href="#cb43-546" aria-hidden="true" tabindex="-1"></a><span class="fu">### DS</span></span>
<span id="cb43-547"><a href="#cb43-547" aria-hidden="true" tabindex="-1"></a><span class="in">```{r DS, message=FALSE, warning=FALSE}</span></span>
<span id="cb43-548"><a href="#cb43-548" aria-hidden="true" tabindex="-1"></a>dsl <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(x) {</span>
<span id="cb43-549"><a href="#cb43-549" aria-hidden="true" tabindex="-1"></a>  e <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DS_edgeR</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, </span>
<span id="cb43-550"><a href="#cb43-550" aria-hidden="true" tabindex="-1"></a>    <span class="st">"cluster_re"</span>, <span class="st">"gene"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb43-551"><a href="#cb43-551" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DS_edgeR"</span>, <span class="at">DADS =</span> <span class="st">"DS"</span>, <span class="at">clustering =</span> <span class="cn">TRUE</span>) </span>
<span id="cb43-552"><a href="#cb43-552" aria-hidden="true" tabindex="-1"></a>  <span class="co">#l &lt;- data.frame(.DS_lemur(sces[[x]])[,c("p_val", "p_adj", "cluster_re")], </span></span>
<span id="cb43-553"><a href="#cb43-553" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  sel = x, das = "DS_lemur", DADS = "DS", clustering = FALSE)</span></span>
<span id="cb43-554"><a href="#cb43-554" aria-hidden="true" tabindex="-1"></a>  m <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">.DS_miloDE</span>(sces[[x]])[,<span class="fu">c</span>(<span class="st">"p_val"</span>, <span class="st">"p_adj"</span>, </span>
<span id="cb43-555"><a href="#cb43-555" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Nhood"</span>, <span class="st">"gene"</span>, <span class="st">"logFC"</span>)], </span>
<span id="cb43-556"><a href="#cb43-556" aria-hidden="true" tabindex="-1"></a>    <span class="at">sel =</span> x, <span class="at">das =</span> <span class="st">"DS_miloDE"</span>, <span class="at">DADS =</span> <span class="st">"DS"</span>, <span class="at">clustering =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb43-557"><a href="#cb43-557" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rename</span>(<span class="st">"cluster_re"</span> <span class="ot">=</span> <span class="st">"Nhood"</span>)</span>
<span id="cb43-558"><a href="#cb43-558" aria-hidden="true" tabindex="-1"></a>  <span class="fu">do.call</span>(rbind, <span class="fu">list</span>(e,m))</span>
<span id="cb43-559"><a href="#cb43-559" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-560"><a href="#cb43-560" aria-hidden="true" tabindex="-1"></a>ds <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, dsl)</span>
<span id="cb43-561"><a href="#cb43-561" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-562"><a href="#cb43-562" aria-hidden="true" tabindex="-1"></a>das <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">list</span>(ds[, <span class="sc">-</span><span class="fu">which</span>(<span class="fu">names</span>(ds) <span class="sc">==</span> <span class="st">"gene"</span>)], da))</span>
<span id="cb43-563"><a href="#cb43-563" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-564"><a href="#cb43-564" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-565"><a href="#cb43-565" aria-hidden="true" tabindex="-1"></a><span class="fu">### Comparison between DA and DS</span></span>
<span id="cb43-566"><a href="#cb43-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-567"><a href="#cb43-567" aria-hidden="true" tabindex="-1"></a>The DA and DS p-value distribution with different feature selection methods.</span>
<span id="cb43-568"><a href="#cb43-568" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-569"><a href="#cb43-569" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Distribution of p values and logFC</span></span>
<span id="cb43-570"><a href="#cb43-570" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-571"><a href="#cb43-571" aria-hidden="true" tabindex="-1"></a>::: {.panel-tabset} </span>
<span id="cb43-572"><a href="#cb43-572" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-573"><a href="#cb43-573" aria-hidden="true" tabindex="-1"></a><span class="fu">#### By selection</span></span>
<span id="cb43-574"><a href="#cb43-574" aria-hidden="true" tabindex="-1"></a><span class="in">```{r sel, message=FALSE, warning=FALSE}</span></span>
<span id="cb43-575"><a href="#cb43-575" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(p_val, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> das, <span class="at">linetype =</span> DADS)) <span class="sc">+</span></span>
<span id="cb43-576"><a href="#cb43-576" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb43-577"><a href="#cb43-577" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span>
<span id="cb43-578"><a href="#cb43-578" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-579"><a href="#cb43-579" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(logFC, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> das, <span class="at">linetype =</span> DADS)) <span class="sc">+</span></span>
<span id="cb43-580"><a href="#cb43-580" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb43-581"><a href="#cb43-581" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span>
<span id="cb43-582"><a href="#cb43-582" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-583"><a href="#cb43-583" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-584"><a href="#cb43-584" aria-hidden="true" tabindex="-1"></a><span class="fu">#### By DAS</span></span>
<span id="cb43-585"><a href="#cb43-585" aria-hidden="true" tabindex="-1"></a><span class="in">```{r DAS, message=FALSE, warning=FALSE}</span></span>
<span id="cb43-586"><a href="#cb43-586" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(p_val, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> sel)) <span class="sc">+</span></span>
<span id="cb43-587"><a href="#cb43-587" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb43-588"><a href="#cb43-588" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> das, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span>
<span id="cb43-589"><a href="#cb43-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-590"><a href="#cb43-590" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(das, <span class="fu">aes</span>(logFC, <span class="fu">after_stat</span>(ndensity), <span class="at">col =</span> sel)) <span class="sc">+</span></span>
<span id="cb43-591"><a href="#cb43-591" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_density</span>(<span class="at">key_glyph =</span> <span class="st">"smooth"</span>) <span class="sc">+</span></span>
<span id="cb43-592"><a href="#cb43-592" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_wrap</span>(<span class="sc">~</span> das, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>)</span>
<span id="cb43-593"><a href="#cb43-593" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-594"><a href="#cb43-594" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-595"><a href="#cb43-595" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb43-596"><a href="#cb43-596" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-597"><a href="#cb43-597" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Clustering-based DA vs Clustering-based DS</span></span>
<span id="cb43-598"><a href="#cb43-598" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-599"><a href="#cb43-599" aria-hidden="true" tabindex="-1"></a>Since the number of clusters between clustering-based DA and DS are the same, we are able to compare the results of *DA_edgeR*, *DA_limma* with the aggregated results of *DS_edgeR* in each cluster. </span>
<span id="cb43-600"><a href="#cb43-600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-601"><a href="#cb43-601" aria-hidden="true" tabindex="-1"></a><span class="in">```{r DADS, fig.height=8, fig.width=12}</span></span>
<span id="cb43-602"><a href="#cb43-602" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> das[das<span class="sc">$</span>clustering <span class="sc">==</span> <span class="cn">TRUE</span>,]</span>
<span id="cb43-603"><a href="#cb43-603" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span></span>
<span id="cb43-604"><a href="#cb43-604" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">color =</span> <span class="fu">case_when</span>(</span>
<span id="cb43-605"><a href="#cb43-605" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"pink"</span>,</span>
<span id="cb43-606"><a href="#cb43-606" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">~</span> <span class="st">"lightblue"</span>,</span>
<span id="cb43-607"><a href="#cb43-607" aria-hidden="true" tabindex="-1"></a>    DADS <span class="sc">==</span> <span class="st">"DA"</span> <span class="sc">~</span> <span class="st">"black"</span>,</span>
<span id="cb43-608"><a href="#cb43-608" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"gray"</span></span>
<span id="cb43-609"><a href="#cb43-609" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb43-610"><a href="#cb43-610" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-611"><a href="#cb43-611" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> p_val, <span class="at">col =</span> DADS, <span class="at">shape =</span> das)) <span class="sc">+</span></span>
<span id="cb43-612"><a href="#cb43-612" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb43-613"><a href="#cb43-613" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb43-614"><a href="#cb43-614" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Pastel1"</span>) <span class="sc">+</span></span>
<span id="cb43-615"><a href="#cb43-615" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> <span class="st">"mean"</span>, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb43-616"><a href="#cb43-616" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.05</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>)</span>
<span id="cb43-617"><a href="#cb43-617" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-618"><a href="#cb43-618" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> logFC, <span class="at">shape =</span> das, <span class="at">col =</span> color)) <span class="sc">+</span></span>
<span id="cb43-619"><a href="#cb43-619" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb43-620"><a href="#cb43-620" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb43-621"><a href="#cb43-621" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_identity</span>() <span class="sc">+</span></span>
<span id="cb43-622"><a href="#cb43-622" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> <span class="st">"mean"</span>, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">8</span>, <span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb43-623"><a href="#cb43-623" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb43-624"><a href="#cb43-624" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>)</span>
<span id="cb43-625"><a href="#cb43-625" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-626"><a href="#cb43-626" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-627"><a href="#cb43-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-628"><a href="#cb43-628" aria-hidden="true" tabindex="-1"></a><span class="fu">#### Clustering-based DA vs Clustering-free DA</span></span>
<span id="cb43-629"><a href="#cb43-629" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-630"><a href="#cb43-630" aria-hidden="true" tabindex="-1"></a>In the <span class="in">`miloR`</span> package, the <span class="in">`annotateNhood`</span> function can be used to annotate each neighborhood with a cluster ID. This is achieved by calculating the maximum proportion of clusters present in that neighborhood.</span>
<span id="cb43-631"><a href="#cb43-631" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-632"><a href="#cb43-632" aria-hidden="true" tabindex="-1"></a>The beeswarm plots are then generated to compare the mean logFC or p-val values over neighborhoods in each cluster with the clustering-based results. These plots provide a visual comparison between the clustering-based results and the mean logFC or p-val values for each neighborhood within the respective clusters.</span>
<span id="cb43-633"><a href="#cb43-633" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-634"><a href="#cb43-634" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-635"><a href="#cb43-635" aria-hidden="true" tabindex="-1"></a><span class="in">```{r milo comparsion, fig.height=8, fig.width=12, warning=FALSE, message=FALSE}</span></span>
<span id="cb43-636"><a href="#cb43-636" aria-hidden="true" tabindex="-1"></a>milos <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(sces), \(i){</span>
<span id="cb43-637"><a href="#cb43-637" aria-hidden="true" tabindex="-1"></a>    x <span class="ot">&lt;-</span> sces[[i]]</span>
<span id="cb43-638"><a href="#cb43-638" aria-hidden="true" tabindex="-1"></a>    cd <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(x))</span>
<span id="cb43-639"><a href="#cb43-639" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">distinct</span>(cd[<span class="fu">c</span>(<span class="st">"sample_id"</span>, <span class="st">"group_id"</span>)])</span>
<span id="cb43-640"><a href="#cb43-640" aria-hidden="true" tabindex="-1"></a>    <span class="fu">rownames</span>(df) <span class="ot">&lt;-</span> df<span class="sc">$</span>sample_id</span>
<span id="cb43-641"><a href="#cb43-641" aria-hidden="true" tabindex="-1"></a>    <span class="co">#x &lt;- Milo(x)</span></span>
<span id="cb43-642"><a href="#cb43-642" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">makeNhoods</span>(<span class="fu">buildGraph</span>(<span class="fu">Milo</span>(x), <span class="at">d =</span> <span class="fu">ncol</span>(<span class="fu">reducedDim</span>(x, <span class="st">"PCA"</span>))))</span>
<span id="cb43-643"><a href="#cb43-643" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">countCells</span>(y, <span class="st">"sample_id"</span>, cd)</span>
<span id="cb43-644"><a href="#cb43-644" aria-hidden="true" tabindex="-1"></a>    y <span class="ot">&lt;-</span> <span class="fu">calcNhoodDistance</span>(y, <span class="at">d =</span> <span class="fu">ncol</span>(<span class="fu">reducedDim</span>(x, <span class="st">"PCA"</span>)))</span>
<span id="cb43-645"><a href="#cb43-645" aria-hidden="true" tabindex="-1"></a>    da <span class="ot">&lt;-</span> <span class="fu">testNhoods</span>(y, <span class="sc">~</span> group_id, df)</span>
<span id="cb43-646"><a href="#cb43-646" aria-hidden="true" tabindex="-1"></a>    res <span class="ot">&lt;-</span> <span class="fu">annotateNhoods</span>(y, <span class="at">da.res =</span> da, <span class="at">coldata_col =</span> <span class="st">"cluster_re"</span>)</span>
<span id="cb43-647"><a href="#cb43-647" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(res, <span class="at">sel =</span> i)</span>
<span id="cb43-648"><a href="#cb43-648" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb43-649"><a href="#cb43-649" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, milos)</span>
<span id="cb43-650"><a href="#cb43-650" aria-hidden="true" tabindex="-1"></a>df1 <span class="ot">&lt;-</span> df1 <span class="sc">%&gt;%</span></span>
<span id="cb43-651"><a href="#cb43-651" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">color =</span> <span class="fu">case_when</span>(</span>
<span id="cb43-652"><a href="#cb43-652" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&gt;</span> <span class="dv">1</span> <span class="sc">~</span> <span class="st">"pink"</span>,</span>
<span id="cb43-653"><a href="#cb43-653" aria-hidden="true" tabindex="-1"></a>    logFC <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">~</span> <span class="st">"lightblue"</span>,</span>
<span id="cb43-654"><a href="#cb43-654" aria-hidden="true" tabindex="-1"></a>    <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"gray"</span></span>
<span id="cb43-655"><a href="#cb43-655" aria-hidden="true" tabindex="-1"></a>  ))</span>
<span id="cb43-656"><a href="#cb43-656" aria-hidden="true" tabindex="-1"></a>df2 <span class="ot">&lt;-</span> das[das<span class="sc">$</span>clustering <span class="sc">==</span> <span class="cn">TRUE</span> <span class="sc">&amp;</span> das<span class="sc">$</span>DADS <span class="sc">==</span> <span class="st">"DA"</span>,]</span>
<span id="cb43-657"><a href="#cb43-657" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-658"><a href="#cb43-658" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df1, <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> logFC, <span class="at">col =</span> color)) <span class="sc">+</span></span>
<span id="cb43-659"><a href="#cb43-659" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> sel, <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb43-660"><a href="#cb43-660" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_quasirandom</span>() <span class="sc">+</span></span>
<span id="cb43-661"><a href="#cb43-661" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_identity</span>() <span class="sc">+</span></span>
<span id="cb43-662"><a href="#cb43-662" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_summary</span>(<span class="at">fun =</span> <span class="st">"mean"</span>, <span class="at">geom =</span> <span class="st">"point"</span>, <span class="at">shape =</span> <span class="dv">8</span>, <span class="at">col =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb43-663"><a href="#cb43-663" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb43-664"><a href="#cb43-664" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"gray"</span>) <span class="sc">+</span> </span>
<span id="cb43-665"><a href="#cb43-665" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> df2, </span>
<span id="cb43-666"><a href="#cb43-666" aria-hidden="true" tabindex="-1"></a>             <span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> cluster_re, <span class="at">y =</span> logFC, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">shape =</span> das))</span>
<span id="cb43-667"><a href="#cb43-667" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb43-668"><a href="#cb43-668" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->



</body></html>